---
title: "PSC article analysis Q2"
output: html_notebook
---

# Custom Functions


```{r}
source("custom_functions.R")
```


# 2. Hypothesis: 2. PSC recurrence is associated with a specific composition of the gut microbiome

**rPSC vs non-rPSC vs Healthy analysis on merged data**

-   Alpha diversity –\> group effect, country effect, interaction effect

-   Beta diversity – PERMANOVA, PCA -\> group effect, country effect,
    interaction effect

-   DAA -\>

    -   Group effect – linDA + MaAsLin2 intersection

    -   Country effect – linDA + MaAsLin2 union

**rPSC vs non-rPSC** –by this comparison, we will find the effect of
recurrence (whether the microbial composition is associated with it)

**non-rPSC vs Healthy** – by this comparison, we will find if the
non-rPSC samples are „closer“ to the healthy samples then rPSC samples
in terms of the microbial composition. In other words, does non-rPSC
samples have „healthy microbiome“?

**rPSC vs Healthy** – by this comparison, we will find how much are the
rPSC samples different from healthy samples.

However, the rPSC and non-rPSC samples are not different at all (see
results). Therefore, we can do EXCLUSIVE LEFT JOIN of differentially
abundant taxa from the analyses (rPSC vs Healthy, non-rPSC vs Healthy)

![imagename](psc_nonrpsc.png)

## Data Import

Importing ASV, taxa and metadata tables for both czech and norway
samples.

### IKEM data

```{r}
asv_tab_ikem <- as.data.frame(fread("data/ikem/asv_table_ikem.csv",check.names = FALSE))
taxa_tab_ikem <- as.data.frame(fread("data/ikem/taxa_table_ikem.csv",check.names = FALSE))
metadata_ikem <- as.data.frame(fread("data/ikem/metadata_ikem.csv",check.names = FALSE))
```

### Norway data

```{r}
asv_tab_norway <- as.data.frame(fread("data/norway/asv_table_norway.csv",check.names = FALSE))
taxa_tab_norway <- as.data.frame(fread("data/norway/taxa_table_norway.csv",check.names = FALSE))
metadata_norway <- as.data.frame(fread("data/norway/metadata_norway.csv",check.names = FALSE))
```

## Merging data

Merging two countries to create whole dataset

```{r}
asv_tab <- merge(asv_tab_ikem,asv_tab_norway,by="SeqID",all=TRUE)
taxa_tab <- merging_taxa_tables(taxa_tab_ikem,taxa_tab_norway)
```

Merging two countries based on the different matrices - Ileum, Colon.

### Terminal ileum

```{r}
ileum_data <- merging_data(asv_tab_1=asv_tab_ikem,
                           asv_tab_2=asv_tab_norway,
                           taxa_tab_1=taxa_tab_ikem,
                           taxa_tab_2=taxa_tab_norway,
                           metadata_1=metadata_ikem,
                           metadata_2=metadata_norway,
                           segment="TI",Q="Q2")

ileum_asv_tab <- ileum_data[[1]]
ileum_taxa_tab <- ileum_data[[2]]
ileum_metadata <- ileum_data[[3]]
```

### Colon

```{r}
colon_data <- merging_data(asv_tab_1=asv_tab_ikem,
                           asv_tab_2=asv_tab_norway,
                           taxa_tab_1=taxa_tab_ikem,
                           taxa_tab_2=taxa_tab_norway,
                           metadata_1=metadata_ikem,
                           metadata_2=metadata_norway,
                           segment="colon",Q="Q2")

colon_asv_tab <- colon_data[[1]]
colon_taxa_tab <- colon_data[[2]]
colon_metadata <- colon_data[[3]]
```

## Analysis - Terminal ileum

### Filtering

Rules: - prevalence \> 5% (per group) - nearZeroVar with default
settings - sequencing depth \> 5000 - taxonomic assignment at least
order

#### Rarefaction Curve

```{r, eval = FALSE}
ps <- construct_phyloseq(ileum_asv_tab,ileum_taxa_tab,ileum_metadata)
rareres <- get_rarecurve(obj=ps, chunks=500)
save(rareres,file = "rarecurves/rarefaction_ileum_H2.Rdata")
```

```{r}
load("rarecurves/rarefaction_ileum.Rdata")

seq_depth_threshold <- 10000
prare <- ggrarecurve(obj=rareres,
                      factorNames="Country",
                      indexNames=c("Observe")) + 
          theme_bw()+
          theme(axis.text=element_text(size=8), panel.grid=element_blank(),
                strip.background = element_rect(colour=NA,fill="grey"),
                strip.text.x = element_text(face="bold")) + 
      geom_vline(xintercept = seq_depth_threshold, linetype="dashed", 
                color = "red") + 
      xlim(0, 20000)
prare
```

Library size

```{r, fig.width=5, fig.height=4, fig.fullwidth=TRUE}
read_counts(ileum_asv_tab, line = c(5000,10000))
```

#### Sequencing depth

```{r}
data_filt <- seq_depth_filtering(ileum_asv_tab,
                                 ileum_taxa_tab,
                                 ileum_metadata,
                                 seq_depth_threshold = 10000)

filt_ileum_asv_tab <- data_filt[[1]]; alpha_ileum_asv_tab <- filt_ileum_asv_tab
filt_ileum_taxa_tab <- data_filt[[2]]; alpha_ileum_taxa_tab <- filt_ileum_taxa_tab
filt_ileum_metadata <- data_filt[[3]]; alpha_ileum_metadata <- filt_ileum_metadata

seq_step <- dim(filt_ileum_asv_tab)[1]
```

Library size

```{r, fig.width=5, fig.height=4, fig.fullwidth=TRUE}
read_counts(filt_ileum_asv_tab,line = c(10000))
```

#### NearZeroVar

```{r}
data_filt <- nearzerovar_filtering(filt_ileum_asv_tab, 
                                   filt_ileum_taxa_tab,
                                   filt_ileum_metadata)
filt_ileum_asv_tab <- data_filt[[1]]
filt_ileum_taxa_tab <- data_filt[[2]]

nearzero_step <- dim(filt_ileum_asv_tab)[1]
```

Library size

```{r, fig.width=5, fig.height=4, fig.fullwidth=TRUE}
read_counts(filt_ileum_asv_tab,line = c(5000,10000))
```
#### Final Counts

```{r}
final_counts_filtering(ileum_asv_tab,
                       filt_ileum_asv_tab,
                       filt_ileum_metadata,
                       seq_step, 0, nearzero_step)
```

### Alpha diversity

```{r}
alpha_div_plots <- list()
```

```{r}
# Construct MPSE object
ileum_mpse <- as.MPSE(construct_phyloseq(alpha_ileum_asv_tab,
                                         alpha_ileum_taxa_tab,
                                         alpha_ileum_metadata))

ileum_mpse %<>% mp_rrarefy(raresize = 10000,seed = 123)

# Calculate alpha diversity - rarefied counts
ileum_mpse %<>% mp_cal_alpha(.abundance=RareAbundance, force=TRUE)
```

#### Plot

```{r}
alpha_data <- data.frame(Observe=ileum_mpse$Observe,
                         Shannon=ileum_mpse$Shannon,
                         Simpson=ileum_mpse$Simpson,
                         Pielou=ileum_mpse$Pielou,
                         Group=ileum_mpse$Group,
                         Country=ileum_mpse$Country)

write.csv(alpha_data,"results/Q2/alpha_data_join_ileum.csv",row.names = FALSE)
```

#### Country plot

```{r, fig.width=10, fig.height=4, fig.fullwidth=TRUE}
labels = data.frame(country=rep(c("CZ","NO"),3),
                   x=c(0.8,1.19,1.8,2.19,2.8,3.19),
                   y=rep(0.2,6),
                   label=rep(c("CZ","NO"),3))

p_boxplot_alpha <- boxplot_subgroups(alpha_data,labels)

# save the results
alpha_div_plots[["Ileum Country"]] <- p_boxplot_alpha

# see the results
p_boxplot_alpha
```


#### Custom plot

```{r}
alpha_data <- alpha_data[,-c(3,4)]
colnames(alpha_data)[1] <- "Richness"
p_A <- alpha_diversity_custom_2(alpha_data,size = 1.5,width = 0.3)

# save the results
alpha_div_plots[["Ileum Custom"]] <- p_A

p_A
```

#### Linear Model

##### Richness

```{r}
alpha_data <- read.csv("results/Q2/alpha_data_join_ileum.csv")
results_model <- pairwise.lm(formula = "Observe ~ Group * Country",
                             factors=alpha_data$Group,
                             data=alpha_data)

# interaction check
if (!is.data.frame(results_model)){
  results_model_observe <- results_model[[1]]
  results_model_observe_emeans <- results_model[[2]]
} else {
  results_model_observe <- results_model
  results_model_observe_emeans <- NA
}

# save the results
pc_observed <- list(); 
pc_observed$terminal_ileum <- results_model_observe

# see the results
results_model_observe
results_model_observe_emeans
```

##### Shannon

```{r}
results_model <- pairwise.lm(formula = "Shannon ~ Group * Country",
                             factors=alpha_data$Group,
                             data=alpha_data)
# interaction check
if (!is.data.frame(results_model)){
  results_model_shannon <- results_model[[1]]
  results_model_shannon_emeans <- results_model[[2]]
} else {
  results_model_shannon <- results_model
  results_model_shannon_emeans <- NA
}


# save the results
pc_shannon <- list(); 
pc_shannon$terminal_ileum <- as.data.frame(results_model_shannon)

# see the results
results_model_shannon
results_model_shannon_emeans
```

##### Simpson

```{r}
results_model <- pairwise.lm(formula = "Simpson ~ Group * Country",
                                     factors=alpha_data$Group,
                                     data=alpha_data)

# interaction check
if (!is.data.frame(results_model)){
  results_model_simpson <- results_model[[1]]
  results_model_simpson_emeans <- results_model[[2]]
} else {
  results_model_simpson <- results_model
  results_model_simpson_emeans <- NA
}


# save the results
pc_simpson <- list(); 
pc_simpson$terminal_ileum <- as.data.frame(results_model_simpson)

# see the results
results_model_simpson
results_model_simpson_emeans
```

##### Pielou

```{r}
results_model <- pairwise.lm(formula = "Pielou ~ Group * Country",
                                     factors=alpha_data$Group,
                                     data=alpha_data)
# interaction check
if (!is.data.frame(results_model)){
  results_model_pielou <- results_model[[1]]
  results_model_pielou_emeans <- results_model[[2]]
} else {
  results_model_pielou <- results_model
  results_model_pielou_emeans <- NA
}


# save the results
pc_pielou <- list(); 
pc_pielou$terminal_ileum <- as.data.frame(results_model_pielou)

# see the results
results_model_pielou
results_model_pielou_emeans
```


### Beta diversity

Calculating Aitchison distance (euclidean distance on clr-transformed
data), both at ASV and genus level.

#### ASV level

##### PERMANOVA

```{r}
# preparing data frame
pairwise_df <- filt_ileum_asv_tab %>% column_to_rownames("SeqID") %>% t()

# main effect
pp_main <- pairwise.adonis(pairwise_df,filt_ileum_metadata$Group,covariate = filt_ileum_metadata$Country, sim.function = "robust.aitchison", p.adjust.m="BH")

# interaction
pp_int <- pairwise.adonis(pairwise_df,filt_ileum_metadata$Group,covariate = filt_ileum_metadata$Country, interaction = TRUE, sim.function = "robust.aitchison", p.adjust.m="BH")

pp_factor <- pp_main[[1]]
pp_cov <- pp_main[[2]]
pp_fac.cov <- pp_int[[3]]

cols <- c("pairs","Df","SumsOfSqs", "F.Model","R2","p.value", "p.adjusted", "sig")
colnames(pp_factor) <- cols; colnames(pp_cov) <- cols; colnames(pp_fac.cov) <- cols; 

# save raw results
pairwise_aitchison_raw <- list()
pairwise_aitchison_raw[["ASV ileum"]] <-rbind(pp_factor,pp_cov,pp_fac.cov)

# see the results
pp_factor
pp_cov
pp_fac.cov

```

Interaction check

```{r}
interaction_sig <- pp_fac.cov$pairs[pp_fac.cov$p.adjusted < 0.05]

if (length(interaction_sig)>0){
 for (i in 1:length(interaction_sig)){
  group1 <- unlist(strsplit(interaction_sig[i],split = " vs "))[1]
  group2 <- unlist(strsplit(interaction_sig[i],split = " vs "))[2]
  group2 <- unlist(strsplit(group2,split = " : "))[1]
  
  result_list <- adonis_postanalysis(x=pairwise_df,
                      factors = filt_ileum_metadata$Group,
                      covariate = filt_ileum_metadata$Country, 
                      group1 = group1,
                      group2 = group2)
  print(result_list)
} 
}

```

##### PCoA

```{r}
pca_plots_list <- list()
```

```{r, eval=FALSE}
ileum_mpse <- as.MPSE(construct_phyloseq(filt_ileum_asv_tab,filt_ileum_taxa_tab,filt_ileum_metadata))
ileum_mpse %<>% mp_cal_dist(.abundance=Abundance, distmethod="robust.aitchison")
ileum_mpse %<>% mp_cal_pcoa(.abundance=Abundance, distmethod="robust.aitchison")

# group-colored
a_plots <- pca_plots(ileum_mpse,by="Group")

# country-colored
ac_plots <- pca_plots(ileum_mpse,by="Country")
```

All in one

```{r, fig.width=10, fig.height=10, fig.fullwidth=TRUE, eval=FALSE}
a_12 <- ggarrange(a_plots[[1]],ac_plots[[1]],ncol=2)
a_13 <- ggarrange(a_plots[[2]],ac_plots[[2]],ncol=2)
a_23 <- ggarrange(a_plots[[3]],ac_plots[[3]],ncol=2)

pca_plots_list <- list()
a <- ggarrange(a_12, a_13, a_23, ncol=1)

# save the results
pca_plots_list[["Q2 Ileum ASV"]] <- a

# see the results
a
```

##### PCoA custom

```{r}
beta_q2 <- pca_plot_custom(filt_ileum_asv_tab,filt_ileum_taxa_tab,filt_ileum_metadata,
                       show_boxplots = TRUE,
                       variable = "Group", size=3, show_legend=TRUE)

# save the results
pca_plots_list[["Q2 Ileum ASV custom"]] <- beta_q2

# see the results
beta_q2

```

#### Genus level

Aggregation, filtering

```{r}
genus_data <- aggregate_taxa(ileum_asv_tab,
                             ileum_taxa_tab,
                             taxonomic_level="Genus",
                             names=TRUE)

filt_data <- filtering_steps(genus_data[[1]],
                             genus_data[[2]],
                             ileum_metadata,
                            seq_depth_threshold=10000)

filt_ileum_genus_tab <- filt_data[[1]]
filt_ileum_genus_taxa <- filt_data[[2]]
filt_ileum_metadata <- filt_data[[3]]
```

##### PERMANOVA

```{r}
pairwise_df <- filt_ileum_genus_tab %>% column_to_rownames("SeqID") %>% t()

# main effect
pp_main <- pairwise.adonis(pairwise_df,filt_ileum_metadata$Group,covariate = filt_ileum_metadata$Country, sim.function = "robust.aitchison", p.adjust.m="BH")

# interaction
pp_int <- pairwise.adonis(pairwise_df,filt_ileum_metadata$Group,covariate = filt_ileum_metadata$Country, interaction = TRUE, sim.function = "robust.aitchison", p.adjust.m="BH")

# tidy the results
pp_factor <- pp_main[[1]]
pp_cov <- pp_main[[2]]
pp_fac.cov <- pp_int[[3]]

cols <- c("pairs","Df","SumsOfSqs", "F.Model","R2","p.value", "p.adjusted", "sig")
colnames(pp_factor) <- cols; colnames(pp_cov) <- cols; colnames(pp_fac.cov) <- cols; 

# save raw results
pairwise_aitchison_raw[["Genus ileum"]] <-rbind(pp_factor,pp_cov,pp_fac.cov)

# see the results
pp_factor
pp_cov
pp_fac.cov
```

Interaction check

```{r}
interaction_sig <- pp_fac.cov$pairs[pp_fac.cov$p.adjusted < 0.05]

if (length(interaction_sig)>0){
for (i in 1:length(interaction_sig)){
  group1 <- unlist(strsplit(interaction_sig[i],split = " vs "))[1]
  group2 <- unlist(strsplit(interaction_sig[i],split = " vs "))[2]
  group2 <- unlist(strsplit(group2,split = " : "))[1]
  
  result_list <- adonis_postanalysis(x=pairwise_df,
                      factors = filt_ileum_metadata$Group,
                      covariate = filt_ileum_metadata$Country, 
                      group1 = group1,
                      group2 = group2)
  print(result_list)
}
}
```

##### PCoA

```{r, eval=FALSE}
ileum_mpse_genus <- as.MPSE(construct_phyloseq(filt_data[[1]],filt_data[[2]],filt_data[[3]]))
ileum_mpse_genus %<>% mp_cal_dist(.abundance=Abundance, distmethod="robust.aitchison")
ileum_mpse_genus %<>% mp_cal_pcoa(.abundance=Abundance, distmethod="robust.aitchison")

# group-colored
g_plots <- pca_plots(ileum_mpse_genus,by="Group")

# country-colored
gc_plots <- pca_plots(ileum_mpse_genus,by="Country")
```

All in one

```{r, fig.width=10, fig.height=10, fig.fullwidth=TRUE, eval=FALSE}
g_12 <- ggarrange(g_plots[[1]],gc_plots[[1]],ncol=2)
g_13 <- ggarrange(g_plots[[2]],gc_plots[[2]],ncol=2)
g_23 <- ggarrange(g_plots[[3]],gc_plots[[3]],ncol=2)

g <- ggarrange(g_12, g_13, g_23, ncol=1)

# save the results
pca_plots_list[["Q2 Ileum Genus"]] <- g

# see the results
g
```

##### PCoA custom

```{r}
beta_q2_genus <- pca_plot_custom(filt_ileum_genus_tab,
                                 filt_ileum_genus_taxa,
                                 filt_ileum_metadata,
                                 show_boxplots = TRUE,
                                 variable = "Group", size=3, show_legend=TRUE)

# save the results
pca_plots_list[["Q2 Ileum Genus custom"]] <- beta_q2_genus

# see the results
beta_q2_genus

```

### Univariate Analysis

#### ASV level

##### rPSC vs non-rPSC

###### linDA

```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
group <- c("non-rPSC","rPSC")

# prepare the data
linda_data <- binomial_prep(ileum_asv_tab,
                            ileum_taxa_tab,
                            ileum_metadata,
                            group, usage="linDA")

filt_ileum_uni_data <- linda_data[[1]]
filt_ileum_uni_taxa <- linda_data[[2]]
filt_ileum_uni_metadata <- linda_data[[3]]

# fit the model
linda.obj <- linda(filt_ileum_uni_data, 
                   filt_ileum_uni_metadata, 
                   formula = '~ Group * Country')

linda.output <- linda.obj$output
linda.output <- linda_renaming(linda.output, group)

# save the results
raw_linda_results <- list()
raw_linda_results[["terminal_ileum"]] <- list()
linda_results <- list()
linda_results[["terminal_ileum"]] <- list()

group1 <- paste0(group[1], " vs ","Group",group[2])
group2 <- paste0(group[1], " , ",group[2], " - CZ vs NO") 
group3 <- paste0(group[1], " vs ","Group",group[2], ":CountryNO")

for (grp in c(group1,group2,group3)){
  raw_linda_results[["terminal_ileum"]][[grp]] <- rawlinda.df(linda.output,
                                                              grp,
                                                              filt_ileum_uni_data,
                                                              filt_ileum_uni_taxa)
  
  linda_results[["terminal_ileum"]][[grp]] <- linda.df(linda.output,
                                                       grp,
                                                       filt_ileum_uni_data,
                                                       filt_ileum_uni_taxa)
}

# summary statistics
raw_linda_results <- binomial_statistics(filt_ileum_uni_data, 
                                         filt_ileum_uni_metadata,
                                         raw_linda_results, 
                                         group=group, 
                                         segment = "terminal_ileum")


# for comparison
wb = createWorkbook()
new_name <- gsub("Group","",paste("linDa il A",group1))
addWorksheet(wb, sheetName = new_name)
linda_df <- rawlinda.df(linda.output,
                        group1,
                        filt_ileum_uni_data,
                        filt_ileum_uni_taxa)

writeData(wb, sheet = new_name, linda_df, rowNames=FALSE)

# see the results
raw_linda_results$terminal_ileum[[group1]]
```


```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
# volcano plot
volcano_1 <- volcano_plot_linda(linda.output, group1,
                                taxa_table = filt_ileum_uni_taxa) +
              ggtitle(paste(group,collapse=" vs "))

volcano_2  <- volcano_plot_linda(linda.output, group2, 
                                 taxa_table = filt_ileum_uni_taxa) + 
              ggtitle("Country effect")

volcano_3 <- volcano_plot_linda(linda.output, group3, 
                                taxa_table = filt_ileum_uni_taxa) + 
              ggtitle("Interaction")

volcano <- ggarrange(volcano_1,volcano_2,volcano_3, ncol=3)

# save the plot
volcano_all_linda <- list()
volcano_all_linda[["terminal_ileum"]] <- list(); i <- 1
volcano_all_linda[["terminal_ileum"]][[i]] <- volcano; i <- i + 1

# see the plot
volcano
```

###### MaAsLin2

```{r, echo=FALSE,results='hide',message=FALSE,warning=FALSE}
fit_data = Maaslin2(
    input_data = filt_ileum_uni_data, 
    input_metadata = filt_ileum_uni_metadata, min_abundance = 0,
    min_prevalence = 0,min_variance = 0,
    output = paste0("results/Q2/",group1), max_significance = 0.05,
    fixed_effects = c('Group', 'Country'),correction = "BH")

```

```{r}
volcano1 <- volcano_plot_maaslin(fit_data,filt_ileum_uni_taxa) + ggtitle(paste(group[1], "vs", group[2]))
volcano2 <- volcano_plot_maaslin(fit_data,filt_ileum_uni_taxa,variable="Country") + ggtitle("Country effect")
volcano <- ggarrange(volcano1,volcano2, ncol=2)

volcano

# for comparison
new_name <- gsub("Group","",paste("Maas il A",group1))
addWorksheet(wb, sheetName = new_name)
maaslin_df <- raw_maaslin.df(fit_data,filt_ileum_uni_data,filt_ileum_uni_taxa)
writeData(wb, sheet = new_name, maaslin_df, rowNames=FALSE)

```

###### Group - Intersection

```{r}
list_intersections <- list()
list_venns <- list()

intersection_results <- group_intersection(group, list_intersections, list_venns,
                                           linda.output, fit_data,
                                           raw_linda_results, 
                                           segment = "terminal_ileum",
                                           level="ASV")

list_intersections <- intersection_results[[1]]
list_venns <- intersection_results[[2]]
venn <- intersection_results[[3]]

# show the results
venn
```

###### Country - Union

```{r, eval=FALSE}
list_country_union <- list()
list_country_union <- country_union(group,linda.output, fit_data,
                                    segment="terminal_ileum",
                                    level="ASV")
```

###### Interaction effect

```{r}
list_interaction_significant <- country_interaction(group,
                                                    linda.output, 
                                                    list_intersections,
                                          filt_ileum_uni_data,
                                          filt_ileum_uni_metadata,
                                          segment="Ileum",
                                          level="ASV")

# see the result
## significant interaction effect
list_interaction_significant[[1]]

## results for czech cohort
list_interaction_significant[[2]]

## results for norwegian cohort
list_interaction_significant[[3]]
```

Removing problematic taxa

```{r}
list_intersections <- removing_interaction_problems(group,
                                                    list_interaction_significant,
                                                    list_intersections,
                                                    segment="Ileum",
                                                    level="ASV")
```

##### rPSC vs healthy

###### linDA

```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
group <- c("healthy","rPSC")

# prepare the data
linda_data <- binomial_prep(ileum_asv_tab,
                            ileum_taxa_tab,
                            ileum_metadata,
                            group, usage="linDA")

filt_ileum_uni_data <- linda_data[[1]]
filt_ileum_uni_taxa <- linda_data[[2]]
filt_ileum_uni_metadata <- linda_data[[3]]

# fit the model
linda.obj <- linda(filt_ileum_uni_data, 
                   filt_ileum_uni_metadata, 
                   formula = '~ Group * Country')

linda.output <- linda.obj$output
linda.output <- linda_renaming(linda.output, group)

# save the results
group1 <- paste0(group[1], " vs ","Group",group[2])
group2 <- paste0(group[1], " , ",group[2], " - CZ vs NO") 
group3 <- paste0(group[1], " vs ","Group",group[2], ":CountryNO")

for (grp in c(group1,group2,group3)){
  raw_linda_results[["terminal_ileum"]][[grp]] <- rawlinda.df(linda.output,
                                                              grp,
                                                              filt_ileum_uni_data,
                                                              filt_ileum_uni_taxa)
  
  linda_results[["terminal_ileum"]][[grp]] <- linda.df(linda.output,
                                                       grp,
                                                       filt_ileum_uni_data,
                                                       filt_ileum_uni_taxa)
}

# summary statistics
raw_linda_results <- binomial_statistics(filt_ileum_uni_data,
                                         filt_ileum_uni_metadata, 
                                         raw_linda_results,
                                         group=group, segment = "terminal_ileum")

# for comparison
wb = createWorkbook()
new_name <- gsub("Group","",paste("linDa il A",group1))
addWorksheet(wb, sheetName = new_name)
linda_df <- rawlinda.df(linda.output,
                        group1,
                        filt_ileum_uni_data,
                        filt_ileum_uni_taxa)
writeData(wb, sheet = new_name, linda_df, rowNames=FALSE)

# see the results
raw_linda_results$terminal_ileum[[group1]]
```

```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
# volcano plot
volcano_1 <- volcano_plot_linda(linda.output, group1,
                                taxa_table = filt_ileum_uni_taxa) +
              ggtitle(paste(group,collapse=" vs "))

volcano_2  <- volcano_plot_linda(linda.output, group2, 
                                 taxa_table = filt_ileum_uni_taxa) + 
              ggtitle("Country effect")

volcano_3 <- volcano_plot_linda(linda.output, group3, 
                                taxa_table = filt_ileum_uni_taxa) + 
              ggtitle("Interaction")

volcano <- ggarrange(volcano_1,volcano_2,volcano_3, ncol=3)

# save the plot
volcano_all_linda[["terminal_ileum"]][[i]] <- volcano; i <- i + 1

# see the plot
volcano
```

###### MaAsLin2

```{r, echo=FALSE,results='hide',message=FALSE,warning=FALSE}
fit_data = Maaslin2(
    input_data = filt_ileum_uni_data, 
    input_metadata = filt_ileum_uni_metadata, min_abundance = 0,
    min_prevalence = 0,min_variance = 0,
    output = paste0("results/Q2/",group1), max_significance = 0.05,
    fixed_effects = c('Group', 'Country'),correction = "BH")
```

```{r}
volcano1 <- volcano_plot_maaslin(fit_data,filt_ileum_uni_taxa) + ggtitle(paste(group[1], "vs", group[2]))
volcano2 <- volcano_plot_maaslin(fit_data,filt_ileum_uni_taxa,variable="Country") + ggtitle("Country effect")
volcano <- ggarrange(volcano1,volcano2, ncol=2)
volcano

# for comparison
new_name <- gsub("Group","",paste("Maas il A",group1))
addWorksheet(wb, sheetName = new_name)
maaslin_df <- raw_maaslin.df(fit_data,filt_ileum_uni_data,filt_ileum_uni_taxa)
writeData(wb, sheet = new_name, maaslin_df, rowNames=FALSE)
```

```{r}
intersection_results <- group_intersection(group, list_intersections, list_venns,
                                           linda.output, fit_data,
                                           raw_linda_results, 
                                           segment = "terminal_ileum",
                                           level="ASV")

list_intersections <- intersection_results[[1]]
list_venns <- intersection_results[[2]]
venn <- intersection_results[[3]]

# show the results
venn
```

###### Country - Union

```{r, eval=FALSE}
list_country_union <- country_union(group,linda.output, fit_data,
                                    segment="terminal_ileum",
                                    level="ASV")
```

###### Interaction effect

```{r}
list_interaction_significant <- country_interaction(group,
                                                    linda.output, 
                                                    list_intersections,
                                          filt_ileum_uni_data,
                                          filt_ileum_uni_metadata,
                                          segment="Ileum",
                                          level="ASV")

# see the result
## significant interaction effect
list_interaction_significant[[1]]

## results for czech cohort
list_interaction_significant[[2]]

## results for norwegian cohort
list_interaction_significant[[3]]
```

Removing problematic taxa

```{r}
list_intersections <- removing_interaction_problems(group,
                                                    list_interaction_significant,
                                                    list_intersections,
                                                    segment="Ileum",
                                                    level="ASV")
```

##### non-rPSC vs healthy

###### linDA

```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
group <- c("healthy","non-rPSC")

# prepare the data
linda_data <- binomial_prep(ileum_asv_tab,
                            ileum_taxa_tab,
                            ileum_metadata,
                            group, usage="linDA")

filt_ileum_uni_data <- linda_data[[1]]
filt_ileum_uni_taxa <- linda_data[[2]]
filt_ileum_uni_metadata <- linda_data[[3]]

# fit the model
linda.obj <- linda(filt_ileum_uni_data, 
                   filt_ileum_uni_metadata,
                   formula = '~ Group * Country')

linda.output <- linda.obj$output
linda.output <- linda_renaming(linda.output, group)

# save the results
group1 <- paste0(group[1], " vs ","Group",group[2])
group2 <- paste0(group[1], " , ",group[2], " - CZ vs NO") 
group3 <- paste0(group[1], " vs ","Group",group[2], ":CountryNO")

for (grp in c(group1,group2,group3)){
  raw_linda_results[["terminal_ileum"]][[grp]] <- rawlinda.df(linda.output,
                                                              grp,
                                                              filt_ileum_uni_data,
                                                              filt_ileum_uni_taxa)
  
  linda_results[["terminal_ileum"]][[grp]] <- linda.df(linda.output,
                                                       grp,
                                                       filt_ileum_uni_data,
                                                       filt_ileum_uni_taxa)
}

# summary statistics
raw_linda_results <- binomial_statistics(filt_ileum_uni_data,
                                         filt_ileum_uni_metadata, 
                                         raw_linda_results, 
                                         group=group, 
                                         segment = "terminal_ileum")

# for comparison
wb = createWorkbook()
new_name <- gsub("Group","",paste("linDa il A",group1))
addWorksheet(wb, sheetName = new_name)
linda_df <- rawlinda.df(linda.output,
                        group1,
                        filt_ileum_uni_data,
                        filt_ileum_uni_taxa)
writeData(wb, sheet = new_name, linda_df, rowNames=FALSE)

# see the results
raw_linda_results$terminal_ileum[[group1]]
```

```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
# volcano plot
volcano_1 <- volcano_plot_linda(linda.output, group1,
                                taxa_table = filt_ileum_uni_taxa) +
              ggtitle(paste(group,collapse=" vs "))

volcano_2  <- volcano_plot_linda(linda.output, group2, 
                                 taxa_table = filt_ileum_uni_taxa) + 
              ggtitle("Country effect")

volcano_3 <- volcano_plot_linda(linda.output, group3, 
                                taxa_table = filt_ileum_uni_taxa) + 
              ggtitle("Interaction")

volcano <- ggarrange(volcano_1,volcano_2,volcano_3, ncol=3)

# save the plot
volcano_all_linda[["terminal_ileum"]][[i]] <- volcano; i <- i + 1

# see the plot
volcano
```

###### MaAsLin2

```{r, echo=FALSE,results='hide',message=FALSE,warning=FALSE}
fit_data = Maaslin2(
    input_data = filt_ileum_uni_data, 
    input_metadata = filt_ileum_uni_metadata, min_abundance = 0,
    min_prevalence = 0,min_variance = 0,
    output = paste0("results/Q2/",group1), max_significance = 0.05,
    fixed_effects = c('Group', 'Country'),correction = "BH")
```

```{r}
volcano1 <- volcano_plot_maaslin(fit_data,filt_ileum_uni_taxa) + ggtitle(paste(group[1], "vs", group[2]))
volcano2 <- volcano_plot_maaslin(fit_data,filt_ileum_uni_taxa,variable="Country") + ggtitle("Country effect")
volcano <- ggarrange(volcano1,volcano2, ncol=2)
volcano

# for comparison
new_name <- gsub("Group","",paste("Maas il A",group1))
addWorksheet(wb, sheetName = new_name)
maaslin_df <- raw_maaslin.df(fit_data,filt_ileum_uni_data,filt_ileum_uni_taxa)
writeData(wb, sheet = new_name, maaslin_df, rowNames=FALSE)
```

###### Group - Intersection

```{r}
intersection_results <- group_intersection(group, 
                                           list_intersections,
                                           list_venns,
                                           linda.output, fit_data,
                                           raw_linda_results, 
                                           segment = "terminal_ileum",
                                           level="ASV")

list_intersections <- intersection_results[[1]]
list_venns <- intersection_results[[2]]
venn <- intersection_results[[3]]

# show the results
venn
```

###### Country - Union

```{r, eval=FALSE}
list_country_union <- country_union(group,linda.output, fit_data,
                                    segment="terminal_ileum",
                                    level="ASV")
```

###### Interaction effect

```{r}
list_interaction_significant <- country_interaction(group,linda.output,
                                                    list_intersections,
                                          filt_ileum_uni_data,
                                          filt_ileum_uni_metadata,
                                          segment="Ileum",
                                          level="ASV")

# see the result
## significant interaction effect
list_interaction_significant[[1]]

## results for czech cohort
list_interaction_significant[[2]]

## results for norwegian cohort
list_interaction_significant[[3]]
```

Removing problematic taxa

```{r}
list_intersections <- removing_interaction_problems(group,
                                                    list_interaction_significant,
                                                    list_intersections,
                                                    segment="Ileum",
                                                    level="ASV")
```

##### Visualizations

Heatmap visualizing the linDA's logFoldChange for taxa with p \< 0.1.

```{r, fig.width=15, fig.height=20, fig.fullwidth=TRUE}
list_heatmap <- list_intersections[grep("Ileum ASV",names(list_intersections),
                                        value=TRUE)]

heatmap_asv_linda <- heatmap_linda(list_heatmap,
                                   ileum_taxa_tab)
heatmap_asv_linda
```

Dot heatmap

```{r, fig.width=5, fig.height=15, fig.fullwidth=TRUE}
dotheatmap_linda <- dot_heatmap_linda(list_heatmap,
                                      ileum_taxa_tab)
dotheatmap_linda
```

##### rPSC effect

**rPSC vs Healthy and non-rPSC vs Healthy LEFT JOIN**

![imagename](psc_nonrpsc.png)

```{r}
A <- list_intersections$`Ileum ASV healthy vs rPSC`
B <- list_intersections$`Ileum ASV healthy vs non-rPSC`

df <- A[!(A$ASV %in% B$ASV),]

rpsc_effect <- list()
rpsc_effect[["Ileum ASV"]] <- df

# see the results
rpsc_effect[["Ileum ASV"]]
```

#### Genus level

Aggregate taxa

```{r}
genus_data <- aggregate_taxa(ileum_asv_tab,
                             ileum_taxa_tab,
                             taxonomic_level = "Genus")

ileum_genus_tab <- genus_data[[1]]
ileum_genus_taxa_tab <- genus_data[[2]]

ileum_genus_asv_taxa_tab <- create_asv_taxa_table(ileum_genus_tab,
                                                  ileum_genus_taxa_tab)
```

##### rPSC vs non-rPSC

###### linDA

```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
group <- c("non-rPSC","rPSC")

# prepare the data
linda_data <- binomial_prep(ileum_genus_tab,
                            ileum_genus_taxa_tab,
                            ileum_metadata,
                            group, usage="linDA")

filt_ileum_uni_data <- linda_data[[1]]
filt_ileum_uni_taxa <- linda_data[[2]]
filt_ileum_uni_metadata <- linda_data[[3]]

# fit the model
linda.obj <- linda(filt_ileum_uni_data,
                   filt_ileum_uni_metadata,
                   formula = '~ Group * Country')

linda.output <- linda.obj$output
linda.output <- linda_renaming(linda.output, group)

# save the results
group1 <- paste0(group[1], " vs ","Group",group[2])
group2 <- paste0(group[1], " , ",group[2], " - CZ vs NO") 
group3 <- paste0(group[1], " vs ","Group",group[2], ":CountryNO")

raw_linda_results_genus <- list();
raw_linda_results_genus[["terminal_ileum"]] <- list()
linda_results_genus <- list(); 
linda_results_genus[["terminal_ileum"]] <- list()

for (grp in c(group1,group2,group3)){
  raw_linda_results_genus[["terminal_ileum"]][[grp]] <- rawlinda.df(linda.output,
                                                                    grp,
                                                                    filt_ileum_uni_data,
                                                                    filt_ileum_uni_taxa)
  linda_results_genus[["terminal_ileum"]][[grp]] <- linda.df(linda.output,
                                                             grp,
                                                             filt_ileum_uni_data,
                                                             filt_ileum_uni_taxa)
}

# summary statistics
raw_linda_results_genus <- binomial_statistics(filt_ileum_uni_data,                     
                                            group=group,
                                            filt_ileum_uni_metadata,
                                            raw_linda_results_genus,
                                            segment = "terminal_ileum")

# for comparison
new_name <- gsub("Group","",paste("linDA il G",group1))
addWorksheet(wb, sheetName = new_name)
linda_df <- rawlinda.df(linda.output,group1,filt_ileum_uni_data,filt_ileum_uni_taxa)
writeData(wb, sheet = new_name, linda_df, rowNames=FALSE)

```

```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
# volcano plot
volcano_1 <- volcano_plot_linda(linda.output, group1, 
                                taxa_table = filt_ileum_uni_taxa) + 
            ggtitle(paste(group,collapse=" vs "))

volcano_2  <- volcano_plot_linda(linda.output, group2, 
                                taxa_table = filt_ileum_uni_taxa) + 
            ggtitle("Country effect") 

volcano_3 <- volcano_plot_linda(linda.output, group3, 
                                taxa_table = filt_ileum_uni_taxa) +
            ggtitle("Interaction")

volcano <- ggarrange(volcano_1,volcano_2,volcano_3, ncol=3)

# save the plot
volcano_all_linda_genus <- list(); 
volcano_all_linda_genus[["terminal_ileum"]] <- list(); i <- 1
volcano_all_linda_genus[["terminal_ileum"]][[i]] <- volcano; i <- i + 1

# see the plot
volcano
```

###### MaAsLin2

```{r, echo=FALSE,results='hide',message=FALSE,warning=FALSE}
fit_data = Maaslin2(
    input_data = filt_ileum_uni_data, 
    input_metadata = filt_ileum_uni_metadata, min_abundance = 0,
    min_prevalence = 0,min_variance = 0,
    output = paste0("results/Q2/G",group1), max_significance = 0.05,
    fixed_effects = c('Group', 'Country'),correction = "BH")

```

```{r}
volcano1 <- volcano_plot_maaslin(fit_data,filt_ileum_uni_taxa) + ggtitle(paste(group[1], "vs", group[2]))
volcano2 <- volcano_plot_maaslin(fit_data,filt_ileum_uni_taxa,variable="Country") + ggtitle("Country effect")
volcano <- ggarrange(volcano1,volcano2, ncol=2)
volcano

# for comparison
new_name <- gsub("Group","",paste("Maas il G",group1))
addWorksheet(wb, sheetName = new_name)
maaslin_df <- raw_maaslin.df(fit_data,filt_ileum_uni_data,filt_ileum_uni_taxa)
writeData(wb, sheet = new_name, maaslin_df, rowNames=FALSE)
```

###### Group - Intersection

```{r}
intersection_results <- group_intersection(group, list_intersections, list_venns,
                                           linda.output, fit_data,
                                           raw_linda_results_genus, 
                                           segment = "terminal_ileum",
                                           level="Genus")

list_intersections <- intersection_results[[1]]
list_venns <- intersection_results[[2]]
venn <- intersection_results[[3]]

# show the results
venn
```

###### Country - Union

```{r, eval=FALSE}
list_country_union <- country_union(group,linda.output, fit_data,
                                    segment="terminal_ileum",
                                    level="Genus")
```

###### Interaction effect

```{r}
list_interaction_significant <- country_interaction(group,linda.output,
                                                    list_intersections,
                                          filt_ileum_uni_data,
                                          filt_ileum_uni_metadata,
                                          segment="Ileum",
                                          level="Genus")

# see the result
## significant interaction effect
list_interaction_significant[[1]]

## results for czech cohort
list_interaction_significant[[2]]

## results for norwegian cohort
list_interaction_significant[[3]]
```
Removing problematic taxa

```{r}
list_intersections <- removing_interaction_problems(group,
                                                    list_interaction_significant,
                                                    list_intersections,
                                                    segment="Ileum",
                                                    level="Genus")
```

##### rPSC vs healthy

###### linDA

```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
#group <- c("rPSC","healthy")
group <- c("healthy","rPSC")

# prepare the data
linda_data <- binomial_prep(ileum_genus_tab,
                            ileum_genus_taxa_tab,
                            ileum_metadata,
                            group, usage="linDA")

filt_ileum_uni_data <- linda_data[[1]]
filt_ileum_uni_taxa <- linda_data[[2]]
filt_ileum_uni_metadata <- linda_data[[3]]

# fit the model
linda.obj <- linda(filt_ileum_uni_data,
                   filt_ileum_uni_metadata,
                   formula = '~ Group * Country')

linda.output <- linda.obj$output
linda.output <- linda_renaming(linda.output, group)

# save the results
group1 <- paste0(group[1], " vs ","Group",group[2])
group2 <- paste0(group[1], " , ",group[2], " - CZ vs NO") 
group3 <- paste0(group[1], " vs ","Group",group[2], ":CountryNO")

raw_linda_results_genus <- list();
raw_linda_results_genus[["terminal_ileum"]] <- list()
linda_results_genus <- list(); 
linda_results_genus[["terminal_ileum"]] <- list()

for (grp in c(group1,group2,group3)){
  raw_linda_results_genus[["terminal_ileum"]][[grp]] <- rawlinda.df(linda.output,
                                                                    grp,
                                                                    filt_ileum_uni_data,
                                                                    filt_ileum_uni_taxa)
  linda_results_genus[["terminal_ileum"]][[grp]] <- linda.df(linda.output,
                                                             grp,
                                                             filt_ileum_uni_data,
                                                             filt_ileum_uni_taxa)
}

# summary statistics
raw_linda_results_genus <- binomial_statistics(filt_ileum_uni_data,                     
                                            group=group,
                                            filt_ileum_uni_metadata,
                                            raw_linda_results_genus,
                                            segment = "terminal_ileum")

# for comparison
new_name <- gsub("Group","",paste("linDA il G",group1))
addWorksheet(wb, sheetName = new_name)
linda_df <- rawlinda.df(linda.output,
                        group1,
                        filt_ileum_uni_data,
                        filt_ileum_uni_taxa)

writeData(wb, sheet = new_name, linda_df, rowNames=FALSE)

```


```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
# volcano plot
volcano_1 <- volcano_plot_linda(linda.output, group1, 
                                taxa_table = filt_ileum_uni_taxa) + 
            ggtitle(paste(group,collapse=" vs "))

volcano_2  <- volcano_plot_linda(linda.output, group2, 
                                taxa_table = filt_ileum_uni_taxa) + 
            ggtitle("Country effect") 

volcano_3 <- volcano_plot_linda(linda.output, group3, 
                                taxa_table = filt_ileum_uni_taxa) +
            ggtitle("Interaction")

volcano <- ggarrange(volcano_1,volcano_2,volcano_3, ncol=3)

# save the plot
volcano_all_linda_genus[["terminal_ileum"]][[i]] <- volcano; i <- i + 1

# see the plot
volcano
```

###### MaAsLin2

```{r, echo=FALSE,results='hide',message=FALSE,warning=FALSE}
fit_data = Maaslin2(
    input_data = filt_ileum_uni_data, 
    input_metadata = filt_ileum_uni_metadata, min_abundance = 0,
    min_prevalence = 0,min_variance = 0,
    output = paste0("results/Q2/G",group1), max_significance = 0.05,
    fixed_effects = c('Group', 'Country'),correction = "BH")
```

```{r}
volcano1 <- volcano_plot_maaslin(fit_data,filt_ileum_uni_taxa) + ggtitle(paste(group[1], "vs", group[2]))
volcano2 <- volcano_plot_maaslin(fit_data,filt_ileum_uni_taxa,variable="Country") + ggtitle("Country effect")
volcano <- ggarrange(volcano1,volcano2, ncol=2)
volcano

# for comparison
new_name <- gsub("Group","",paste("Maas il G",group1))
addWorksheet(wb, sheetName = new_name)
maaslin_df <- raw_maaslin.df(fit_data,filt_ileum_uni_data,filt_ileum_uni_taxa)
writeData(wb, sheet = new_name, maaslin_df, rowNames=FALSE)
```

###### Group - Intersection

```{r}
intersection_results <- group_intersection(group, list_intersections, list_venns,
                                           linda.output, fit_data,
                                           raw_linda_results_genus, 
                                           segment = "terminal_ileum",
                                           level="Genus")

list_intersections <- intersection_results[[1]]
list_venns <- intersection_results[[2]]
venn <- intersection_results[[3]]

# show the results
venn
```

###### Country - Union

```{r eval=FALSE}
list_country_union <- country_union(group,linda.output, fit_data,
                                    segment="terminal_ileum",
                                    level="Genus")
```

###### Interaction effect

```{r}
list_interaction_significant <- country_interaction(group,linda.output, list_intersections,
                                          filt_ileum_uni_data,
                                          filt_ileum_uni_metadata,
                                          segment="Ileum",
                                          level="Genus")

# see the result
## significant interaction effect
list_interaction_significant[[1]]

## results for czech cohort
list_interaction_significant[[2]]

## results for norwegian cohort
list_interaction_significant[[3]]
```
Removing problematic taxa

```{r}
list_intersections <- removing_interaction_problems(group,
                                                    list_interaction_significant,
                                                    list_intersections,
                                                    segment="Ileum",
                                                    level="Genus")
```

##### non-rPSC vs healthy

###### linDA

```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
#group <- c("non-rPSC","healthy")
group <- c("healthy","non-rPSC")

# prepare the data
linda_data <- binomial_prep(ileum_genus_tab,
                            ileum_genus_taxa_tab,
                            ileum_metadata,
                            group, usage="linDA")

filt_ileum_uni_data <- linda_data[[1]]
filt_ileum_uni_taxa <- linda_data[[2]]
filt_ileum_uni_metadata <- linda_data[[3]]

# fit the model
linda.obj <- linda(filt_ileum_uni_data,
                   filt_ileum_uni_metadata,
                   formula = '~ Group * Country')

linda.output <- linda.obj$output
linda.output <- linda_renaming(linda.output, group)

# save the results
group1 <- paste0(group[1], " vs ","Group",group[2])
group2 <- paste0(group[1], " , ",group[2], " - CZ vs NO") 
group3 <- paste0(group[1], " vs ","Group",group[2], ":CountryNO")

raw_linda_results_genus <- list();
raw_linda_results_genus[["terminal_ileum"]] <- list()
linda_results_genus <- list(); 
linda_results_genus[["terminal_ileum"]] <- list()

for (grp in c(group1,group2,group3)){
  raw_linda_results_genus[["terminal_ileum"]][[grp]] <- rawlinda.df(linda.output,
                                                                    grp,
                                                                    filt_ileum_uni_data,
                                                                    filt_ileum_uni_taxa)
  linda_results_genus[["terminal_ileum"]][[grp]] <- linda.df(linda.output,
                                                             grp,
                                                             filt_ileum_uni_data,
                                                             filt_ileum_uni_taxa)
}

# summary statistics
raw_linda_results_genus <- binomial_statistics(filt_ileum_uni_data,                     
                                            group=group,
                                            filt_ileum_uni_metadata,
                                            raw_linda_results_genus,
                                            segment = "terminal_ileum")

# for comparison
new_name <- gsub("Group","",paste("linDA il G",group1))
addWorksheet(wb, sheetName = new_name)

linda_df <- rawlinda.df(linda.output,
                        group1,
                        filt_ileum_uni_data,
                        filt_ileum_uni_taxa)

writeData(wb, sheet = new_name, linda_df, rowNames=FALSE)

```


```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
# volcano plot
volcano_1 <- volcano_plot_linda(linda.output, group1, 
                                taxa_table = filt_ileum_uni_taxa) + 
            ggtitle(paste(group,collapse=" vs "))

volcano_2  <- volcano_plot_linda(linda.output, group2, 
                                taxa_table = filt_ileum_uni_taxa) + 
            ggtitle("Country effect") 

volcano_3 <- volcano_plot_linda(linda.output, group3, 
                                taxa_table = filt_ileum_uni_taxa) +
            ggtitle("Interaction")

volcano <- ggarrange(volcano_1,volcano_2,volcano_3, ncol=3)

# save the plot
volcano_all_linda_genus[["terminal_ileum"]][[i]] <- volcano; i <- i + 1

# see the plot
volcano
```

###### MaAsLin2

```{r, echo=FALSE,results='hide',message=FALSE,warning=FALSE}
fit_data = Maaslin2(
    input_data = filt_ileum_uni_data, 
    input_metadata = filt_ileum_uni_metadata, min_abundance = 0,
    min_prevalence = 0,min_variance = 0,
    output = paste0("results/Q2/G",group1), max_significance = 0.05,
    fixed_effects = c('Group', 'Country'),correction = "BH")
```

```{r}
volcano1 <- volcano_plot_maaslin(fit_data,filt_ileum_uni_taxa) + ggtitle(paste(group[1], "vs", group[2]))
volcano2 <- volcano_plot_maaslin(fit_data,filt_ileum_uni_taxa,variable="Country") + ggtitle("Country effect")
volcano <- ggarrange(volcano1,volcano2, ncol=2)
volcano

# for comparison
new_name <- gsub("Group","",paste("Maas il G",group1))
addWorksheet(wb, sheetName = new_name)
maaslin_df <- raw_maaslin.df(fit_data,filt_ileum_uni_data,filt_ileum_uni_taxa)
writeData(wb, sheet = new_name, maaslin_df, rowNames=FALSE)
```

###### Group - Intersection

```{r}
intersection_results <- group_intersection(group, list_intersections, list_venns,
                                           linda.output, fit_data,
                                           raw_linda_results_genus, 
                                           segment = "terminal_ileum",
                                           level="Genus")

list_intersections <- intersection_results[[1]]
list_venns <- intersection_results[[2]]
venn <- intersection_results[[3]]

# show the results
venn
```

###### Country - Union

```{r, eval=FALSE}
list_country_union <- country_union(group,linda.output, fit_data,
                                    segment="terminal_ileum",
                                    level="Genus")
```

###### Interaction effect

```{r}
list_interaction_significant <- country_interaction(group,linda.output,
                                                    list_intersections,
                                          filt_ileum_uni_data,
                                          filt_ileum_uni_metadata,
                                          segment="Ileum",
                                          level="Genus")

# see the result
## significant interaction effect
list_interaction_significant[[1]]

## results for czech cohort
list_interaction_significant[[2]]

## results for norwegian cohort
list_interaction_significant[[3]]
```

Removing problematic taxa

```{r}
list_intersections <- removing_interaction_problems(group,
                                                    list_interaction_significant,
                                                    list_intersections,
                                                    segment="Ileum",
                                                    level="Genus")
```

##### Visualizations

Heatmap visualizing the linDA's logFoldChange for taxa with p \< 0.1.

```{r, fig.width=10, fig.height=10, fig.fullwidth=TRUE}
list_heatmap <- list_intersections[grep("Ileum Genus",names(list_intersections),value=TRUE)]
heatmap_asv_linda <- heatmap_linda(list_heatmap,ileum_taxa_tab)
heatmap_asv_linda
```

Dot heatmap

```{r, fig.width=5, fig.height=17, fig.fullwidth=TRUE}
dotheatmap_linda <- dot_heatmap_linda(list_heatmap,ileum_taxa_tab)
dotheatmap_linda
```

##### rPSC effect

**pre_LTx vs Healthy and Post_LTx vs Healthy intersection**

![imagename](pre_post.png)

```{r}
A <- list_intersections$`Ileum Genus healthy vs rPSC`
B <- list_intersections$`Ileum Genus healthy vs non-rPSC`

df <- A[!(A$ASV %in% B$ASV),]

rpsc_effect[["Ileum Genus"]] <- df

# see the results
rpsc_effect[["Ileum Genus"]]
```

### ElasticNet - merged cohorts

#### ASV level

##### rPSC vs non-rPSC

```{r}
group <- c("rPSC","non-rPSC")
model_name <- paste(group[1],"vs",group[2],"ASV terminal_ileum")
  
# prepare the data
filt_ileum_uni_data <- binomial_prep(ileum_asv_tab,
                                     ileum_taxa_tab,
                                     ileum_metadata,
                                     group, usage="glmnet")

# fit the model
enet_model <- glmnet_binomial(filt_ileum_uni_data,
                              sample_method = "atypboot",
                              outcome="Group", 
                              N=500,
                              reuse=TRUE,
                              file=model_name,
                              Q="Q2")

# ROC
roc_c <- roc_curve(enet_model, group)

# save the results
models_summ <- list()
models_cm <- list()
betas <- list()
roc_cs <- list()

models_summ[[paste(group[1],"vs",group[2],"ASV terminal_ileum")]] <- enet_model$model_summary
models_cm[[paste(group[1],"vs",group[2],"ASV terminal_ileum")]] <- enet_model$conf_matrices$original
roc_cs[[paste(group[1],"vs",group[2],"ASV terminal_ileum")]] <- enet_model$kfold_rocobjs
betas[[paste(group[1],"vs",group[2],"ASV terminal_ileum")]] <- as.matrix(enet_model$betas)

# see the results
enet_model$model_summary %>% t()
enet_model$conf_matrices
enet_model$plot
roc_c
```

##### rPSC vs healthy

```{r}
group <- c("rPSC","healthy")
model_name <- paste(group[1],"vs",group[2],"ASV terminal_ileum")

# prepare the data
filt_ileum_uni_data <- binomial_prep(ileum_asv_tab,
                                     ileum_taxa_tab,
                                     ileum_metadata,
                                     group, usage="glmnet")

# fit the model
enet_model <- glmnet_binomial(filt_ileum_uni_data,
                              sample_method = "atypboot",
                              outcome="Group", 
                              N=500,
                              reuse=TRUE,
                              file=model_name,
                              Q="Q2")

# ROC
roc_c <- roc_curve(enet_model, group)

# save the results
models_summ[[paste(group[1],"vs",group[2],"ASV terminal_ileum")]] <- enet_model$model_summary
models_cm[[paste(group[1],"vs",group[2],"ASV terminal_ileum")]] <- enet_model$conf_matrices$original
roc_cs[[paste(group[1],"vs",group[2],"ASV terminal_ileum")]] <- enet_model$kfold_rocobjs
betas[[paste(group[1],"vs",group[2],"ASV terminal_ileum")]] <- as.matrix(enet_model$betas)

# see the results
enet_model$model_summary %>% t()
enet_model$conf_matrices
enet_model$plot
roc_c
```

##### non-rPSC vs healthy

```{r}
group <- c("non-rPSC","healthy")
model_name <- paste(group[1],"vs",group[2],"ASV terminal_ileum")

# prepare the data
filt_ileum_uni_data <- binomial_prep(ileum_asv_tab,
                                     ileum_taxa_tab,
                                     ileum_metadata,
                                     group, usage="glmnet")

# fit the model
enet_model <- glmnet_binomial(filt_ileum_uni_data,
                              sample_method = "atypboot",
                              outcome="Group", 
                              N=500,
                              reuse=TRUE,
                              file=model_name,
                              Q="Q2")

# ROC
roc_c <- roc_curve(enet_model, group)

# save the results
models_summ[[paste(group[1],"vs",group[2],"ASV terminal_ileum")]] <- enet_model$model_summary
models_cm[[paste(group[1],"vs",group[2],"ASV terminal_ileum")]] <- enet_model$conf_matrices$original
roc_cs[[paste(group[1],"vs",group[2],"ASV terminal_ileum")]] <- enet_model$kfold_rocobjs
betas[[paste(group[1],"vs",group[2],"ASV terminal_ileum")]] <- as.matrix(enet_model$betas)

# see the results
enet_model$model_summary %>% t()
enet_model$conf_matrices
enet_model$plot
roc_c
```

##### rPSC effect

```{r}
model_name <- paste("rPSC effect","ASV terminal_ileum")

# prepare the data
filt_ileum_uni_data <- binomial_prep_psc_effect(ileum_asv_tab,
                                                ileum_taxa_tab, 
                                                ileum_metadata,
                                                rpsc_effect$`Ileum ASV`,
                                                usage="glmnet")

# fit the model
enet_model <- glmnet_binomial(filt_ileum_uni_data,
                              sample_method = "atypboot",
                              outcome="Group", 
                              N=500,
                              reuse=TRUE,
                              file=model_name,
                              Q="Q2")

# save the results
models_summ[[paste("rPSC effect","ASV terminal_ileum")]] <- enet_model$model_summary
models_cm[[paste("rPSC effect","ASV terminal_ileum")]] <- enet_model$conf_matrices$original
roc_c <- roc_curve(enet_model, group)
roc_cs[[paste("rPSC effect","ASV terminal_ileum")]] <- enet_model$kfold_rocobjs

# see the results
enet_model$model_summary %>% t()
enet_model$conf_matrices
enet_model$plot

roc_c
```

#### Genus level

Aggregate taxa

```{r}
genus_data <- aggregate_taxa(ileum_asv_tab,
                             ileum_taxa_tab,
                             taxonomic_level = "Genus")

ileum_genus_tab <- genus_data[[1]]
ileum_genus_taxa_tab <- genus_data[[2]]
```

##### rPSC vs non-rPSC

```{r}
group <- c("rPSC","non-rPSC")
model_name <- paste(group[1],"vs",group[2],"Genus terminal_ileum")

# prepare the data
filt_ileum_uni_data <- binomial_prep(ileum_genus_tab,
                                     ileum_genus_taxa_tab,
                                     ileum_metadata,group, usage="glmnet")

# fit the model
enet_model <- glmnet_binomial(filt_ileum_uni_data,
                              sample_method = "atypboot",
                              outcome="Group", 
                              N=500,
                              reuse=TRUE,
                              file=model_name,
                              Q="Q2")

# ROC
roc_c <- roc_curve(enet_model, group)

# save the results
models_summ[[paste(group[1],"vs",group[2],"Genus terminal_ileum")]] <- enet_model$model_summary
models_cm[[paste(group[1],"vs",group[2],"Genus terminal_ileum")]] <- enet_model$conf_matrices$original
roc_cs[[paste(group[1],"vs",group[2],"Genus terminal_ileum")]] <- enet_model$kfold_rocobjs
betas[[paste(group[1],"vs",group[2],"Genus terminal_ileum")]] <- as.matrix(enet_model$betas)

# see the results
enet_model$model_summary %>% t()
enet_model$conf_matrices
enet_model$plot

roc_c
```

##### rPSC vs healthy

```{r}
group <- c("rPSC","healthy")
model_name <- paste(group[1],"vs",group[2],"Genus terminal_ileum")

# prepare the data
filt_ileum_uni_data <- binomial_prep(ileum_genus_tab,
                                     ileum_genus_taxa_tab,
                                     ileum_metadata,group, usage="glmnet")

# fit the model
enet_model <- glmnet_binomial(filt_ileum_uni_data,
                              sample_method = "atypboot",
                              outcome="Group", N=500,
                              reuse=TRUE,
                              file=model_name,
                              Q="Q2")

# ROC
roc_c <- roc_curve(enet_model, group)

# save the results
models_summ[[paste(group[1],"vs",group[2],"Genus terminal_ileum")]] <- enet_model$model_summary
models_cm[[paste(group[1],"vs",group[2],"Genus terminal_ileum")]] <- enet_model$conf_matrices$original
roc_cs[[paste(group[1],"vs",group[2],"Genus terminal_ileum")]] <- enet_model$kfold_rocobjs
betas[[paste(group[1],"vs",group[2],"Genus terminal_ileum")]] <- as.matrix(enet_model$betas)

# see the results
enet_model$model_summary %>% t()
enet_model$conf_matrices
enet_model$plot

roc_c
```

##### non-rPSC vs healthy

```{r}
group <- c("non-rPSC","healthy")
model_name <- paste(group[1],"vs",group[2],"Genus terminal_ileum")

# prepare the data
filt_ileum_uni_data <- binomial_prep(ileum_genus_tab,
                                     ileum_genus_taxa_tab,
                                     ileum_metadata,group, 
                                     usage="glmnet")

# fit the model
enet_model <- glmnet_binomial(filt_ileum_uni_data,
                              sample_method = "atypboot",
                              outcome="Group", 
                              N=500,
                              reuse=TRUE,
                              file=model_name,
                              Q="Q2")

# ROC
roc_c <- roc_curve(enet_model, group)

# save the results
models_summ[[paste(group[1],"vs",group[2],"Genus terminal_ileum")]] <- enet_model$model_summary
models_cm[[paste(group[1],"vs",group[2],"Genus terminal_ileum")]] <- enet_model$conf_matrices$original
roc_cs[[paste(group[1],"vs",group[2],"Genus terminal_ileum")]] <- enet_model$kfold_rocobjs
betas[[paste(group[1],"vs",group[2],"Genus terminal_ileum")]] <- as.matrix(enet_model$betas)

# see the results
enet_model$model_summary %>% t()
enet_model$conf_matrices
enet_model$plot

roc_c
```

##### rPSC effect

```{r}
model_name <- paste("rPSC effect","Genus terminal_ileum")

# prepare the data
filt_ileum_uni_data <- binomial_prep_psc_effect(ileum_genus_tab,
                                                ileum_genus_taxa_tab, 
                                     ileum_metadata,
                                     rpsc_effect$`Ileum Genus`,
                                     usage="glmnet")

# fit the model
enet_model <- glmnet_binomial(filt_ileum_uni_data,
                              sample_method = "atypboot",
                              outcome="Group", N=500,
                              reuse=TRUE,
                              file=model_name,
                              Q="Q2")

# save the results
models_summ[[paste("rPSC effect","Genus terminal_ileum")]] <- enet_model$model_summary
models_cm[[paste("rPSC effect","Genus terminal_ileum")]] <- enet_model$conf_matrices$original
roc_c <- roc_curve(enet_model, group)
roc_cs[[paste("rPSC effect","Genus terminal_ileum")]] <- enet_model$kfold_rocobjs

# see the results
enet_model$model_summary %>% t()
enet_model$conf_matrices
enet_model$plot

roc_c
```

### Results overview

#### Alpha diversity

```{r}
pc_observed$terminal_ileum
pc_shannon$terminal_ileum
pc_simpson$terminal_ileum
pc_pielou$terminal_ileum

write.xlsx(x = list(pc_observed$terminal_ileum,
                    pc_shannon$terminal_ileum,
                    pc_simpson$terminal_ileum,
                    pc_pielou$terminal_ileum), 
           file = "results/Q2/alpha_diversity_results_terminal_ileum.xlsx")
```

```{r, fig.width=10, fig.height=4, fig.fullwidth=TRUE}
alpha_div_plots[["Ileum Country"]]
```

```{r, fig.width=10, fig.height=4, fig.fullwidth=TRUE}
alpha_div_plots[["Ileum Custom"]]
```

#### Beta diversity

**Raw results**

```{r}
pairwise_aitchison_raw$`ASV ileum`
pairwise_aitchison_raw$`Genus ileum`
```

**PCA at ASV level**


```{r, fig.width=7, fig.height=5, fig.fullwidth=TRUE}
pca_plots_list$`Q2 Ileum ASV custom`
```

**PCA at Genus level**

```{r, fig.width=10, fig.height=10, fig.fullwidth=TRUE}
pca_plots_list$`Q2 Ileum Genus`
```

```{r, fig.width=7, fig.height=5, fig.fullwidth=TRUE}
pca_plots_list$`Q2 Ileum Genus custom`
```

#### Univariate analysis

**Number of significant taxa**

```{r}
cbind(as.data.frame(lapply(list_intersections,nrow)),
      as.data.frame(lapply(rpsc_effect,nrow))) %>% t() %>% 
  `colnames<-`("Count") %>% 
  `rownames<-`(c(names(list_intersections),"rPSC effect ASV","rPSC effect Genus"))
```

```{r}
# save the results
write.xlsx(x = list_intersections[grep("Ileum",names(list_intersections),value=TRUE)], 
           file = "results/Q2/DAA_terminal_ileum.xlsx")

write.xlsx(x = rpsc_effect[grep("Ileum",names(rpsc_effect),value=TRUE)], 
           file = "results/Q2/rPSC_effect_terminal_ileum.xlsx")

saveWorkbook(wb,"results/Q2/DAA_linDA_Maaslin_terminal_ileum.xlsx", overwrite = TRUE)
```

**Counts**

```{r}
univar_list <- univariate_statistics(list_intersections,
                                     rpsc_effect,
                                     ileum_genus_asv_taxa_tab)

univar_df <- univar_list[[1]]
wb <- univar_list[[2]]

# save the results
saveWorkbook(wb,"results/Q2/DAA_final_terminal_ileum.xlsx", overwrite = TRUE)

# see the results
univar_df
```

#### ElasticNet

Summary

```{r}
models_summ_df <- do.call(
  rbind, 
  models_summ[grep("terminal_ileum",names(models_summ),value = TRUE)])

# save the results
write.xlsx(list(binomial=models_summ_df),
           "results/Q2/elastic_net_summary_terminal_ileum.xlsx", rowNames=TRUE)

# see the results
models_summ_df
```

**ROC - ASV level**

```{r}
roc_curve_all_custom(roc_cs[c(1:4)], "Q2")
```

**ROC - Genus level**

```{r}
roc_curve_all_custom(roc_cs[c(5:8)],"Q2")
```

## Analysis - Colon

### Filtering

Rules: - prevalence \> 5% (per group) - nearZeroVar with default
settings - sequencing depth \> 5000 - taxonomic assignment at least
order

#### Rarefaction Curve

```{r, eval = FALSE}
ps <- construct_phyloseq(colon_asv_tab,colon_taxa_tab,colon_metadata)
rareres <- get_rarecurve(obj=ps, chunks=500)
save(rareres,file = "rarecurves/rarefaction_colon.Rdata")
```

```{r}
load("rarecurves/rarefaction_colon.Rdata")

seq_depth_threshold <- 10000
prare <- ggrarecurve(obj=rareres,
                      factorNames="Country",
                      indexNames=c("Observe")) + 
          theme_bw()+
          theme(axis.text=element_text(size=8), panel.grid=element_blank(),
                strip.background = element_rect(colour=NA,fill="grey"),
                strip.text.x = element_text(face="bold")) + 
      geom_vline(xintercept = seq_depth_threshold, linetype="dashed", 
                color = "red") + 
      xlim(0, 20000)
prare
```

Library size

```{r, fig.width=5, fig.height=4, fig.fullwidth=TRUE}
read_counts(colon_asv_tab, line = c(5000,10000))
```

#### Sequencing depth

```{r}
data_filt <- seq_depth_filtering(colon_asv_tab,
                                 colon_taxa_tab,
                                 colon_metadata,
                                 seq_depth_threshold = 10000)

filt_colon_asv_tab <- data_filt[[1]]; alpha_colon_asv_tab <- filt_colon_asv_tab
filt_colon_taxa_tab <- data_filt[[2]]; alpha_colon_taxa_tab <- filt_colon_taxa_tab
filt_colon_metadata <- data_filt[[3]]; alpha_colon_metadata <- filt_colon_metadata

seq_step <- dim(filt_colon_asv_tab)[1]
```

Library size

```{r, fig.width=5, fig.height=4, fig.fullwidth=TRUE}
read_counts(filt_colon_asv_tab,line = c(10000))
```

#### NearZeroVar

```{r}
data_filt <- nearzerovar_filtering(filt_colon_asv_tab, 
                                   filt_colon_taxa_tab,
                                   filt_colon_metadata)
filt_colon_asv_tab <- data_filt[[1]]
filt_colon_taxa_tab <- data_filt[[2]]

nearzero_step <- dim(filt_colon_asv_tab)[1]
```

Library size

```{r, fig.width=5, fig.height=4, fig.fullwidth=TRUE}
read_counts(filt_colon_asv_tab,line = c(5000,10000))
```

#### Final Counts

```{r}
final_counts_filtering(colon_asv_tab,
                       filt_colon_asv_tab,
                       filt_colon_metadata,
                       seq_step, 0, nearzero_step)
```

### Alpha diversity

```{r}
# Construct MPSE object
colon_mpse <- as.MPSE(construct_phyloseq(alpha_colon_asv_tab,
                                         alpha_colon_taxa_tab,
                                         alpha_colon_metadata))

colon_mpse %<>% mp_rrarefy(raresize = 10000,seed = 123)

# Calculate alpha diversity - rarefied counts
colon_mpse %<>% mp_cal_alpha(.abundance=RareAbundance, force=TRUE)
```

#### Plots

```{r}
alpha_data <- data.frame(Observe=colon_mpse$Observe,
                         Shannon=colon_mpse$Shannon,
                         Simpson=colon_mpse$Simpson,
                         Pielou=colon_mpse$Pielou,
                         Group=colon_mpse$Group,
                         Country=colon_mpse$Country,
                         Patient=colon_mpse$Patient)

write.csv(alpha_data,"results/Q2/alpha_data_join_colon.csv",row.names = FALSE)
```

#### Country plot

```{r, fig.width=10, fig.height=4, fig.fullwidth=TRUE}
# preparing data frame
labels = data.frame(country=rep(c("CZ","NO"),3),
                   x=c(0.8,1.19,1.8,2.19,2.8,3.19),
                   y=rep(0.2,6),
                   label=rep(c("CZ","NO"),3))

p_boxplot_alpha <- boxplot_subgroups(alpha_data,labels)

# save the results
alpha_div_plots[["Colon Country"]] <- p_boxplot_alpha

# see the results
p_boxplot_alpha
```

#### Custom plot

```{r}
alpha_data <- alpha_data[,-c(3,4)]
colnames(alpha_data)[1] <- "Richness"
p_B <- alpha_diversity_custom_2(alpha_data,size = 1,width = 0.3)

# save the results
alpha_div_plots[["Colon Custom"]] <- p_B

p_B
```

#### Linear Model

##### Richness

```{r}
alpha_data <- read.csv("results/Q2/alpha_data_join_colon.csv")
results_model <- pairwise.lmer(
  formula = "Observe ~ Group * Country + (1|Patient)",
  factors=alpha_data$Group,
  data=alpha_data)

# interaction check
if (!is.data.frame(results_model)){
  results_model_observe <- results_model[[1]]
  results_model_observe_detailed <- results_model[[2]]
} else {
  results_model_observe <- results_model
  results_model_observe_detailed <- NA
}

# save the results
pc_observed$colon <- results_model_observe

# see the results
results_model_observe
results_model_observe_detailed
```

##### Shannon

```{r}
results_model <- pairwise.lmer(
  formula = "Shannon ~ Group * Country + (1|Patient)",
  factors=alpha_data$Group,
  data=alpha_data)

# interaction check
if (!is.data.frame(results_model)){
  results_model_shannon <- results_model[[1]]
  results_model_shannon_detailed <- results_model[[2]]
} else {
  results_model_shannon <- results_model
  results_model_shannon_detailed <- NA
}

# save the results
pc_shannon$colon <- as.data.frame(results_model_shannon)

# see the results
results_model_shannon
results_model_shannon_detailed
```

##### Simpson

```{r}
results_model <- pairwise.lmer(
  formula = "Simpson ~ Group * Country + (1|Patient)",
  factors=alpha_data$Group,
  data=alpha_data)

# interaction check
if (!is.data.frame(results_model)){
  results_model_simpson <- results_model[[1]]
  results_model_simpson_detailed <- results_model[[2]]
} else {
  results_model_simpson <- results_model
  results_model_simpson_detailed <- NA
}


# save the results
pc_simpson$colon <- as.data.frame(results_model_simpson)

# see the results
results_model_simpson
results_model_simpson_detailed
```

##### Pielou

```{r}
results_model <- pairwise.lmer(
  formula = "Pielou ~ Group * Country + (1|Patient)",
  factors=alpha_data$Group,
  data=alpha_data)

# interaction check
if (!is.data.frame(results_model)){
  results_model_pielou <- results_model[[1]]
  results_model_pielou_detailed <- results_model[[2]]
} else {
  results_model_pielou <- results_model
  results_model_pielou_detailed <- NA
}

# save the results
pc_pielou$colon <- as.data.frame(results_model_pielou)

# see the results
results_model_pielou
results_model_pielou_detailed
```

### Beta diversity

Calculating Aitchison distance (euclidean distance on clr-transformed
data), both at ASV and genus level.

#### ASV level

##### PERMANOVA

```{r}
# preparing data frame
pairwise_df <- filt_colon_asv_tab %>% column_to_rownames("SeqID") %>% t()

# main effect
pp_main <- pairwise.adonis(pairwise_df,filt_colon_metadata$Group,
                           covariate = filt_colon_metadata$Country, 
                           sim.function = "robust.aitchison", 
                           p.adjust.m="BH",
                           patients = filt_colon_metadata$Patient)

# interaction
pp_int <- pairwise.adonis(pairwise_df,filt_colon_metadata$Group,
                          covariate = filt_colon_metadata$Country, 
                          interaction = TRUE, 
                          sim.function = "robust.aitchison", 
                          p.adjust.m="BH",
                          patients = filt_colon_metadata$Patient)

pp_factor <- pp_main[[1]]
pp_cov <- pp_main[[2]]
pp_fac.cov <- pp_int[[3]]

cols <- c("pairs","Df","SumsOfSqs", "F.Model","R2","p.value", "p.adjusted", "sig")
colnames(pp_factor) <- cols; colnames(pp_cov) <- cols; colnames(pp_fac.cov) <- cols; 

# save raw results
pairwise_aitchison_raw[["ASV colon"]] <-rbind(pp_factor,pp_cov,pp_fac.cov)

# see the results
pp_factor
pp_cov
pp_fac.cov

```

Interaction check

```{r}
interaction_sig <- pp_fac.cov$pairs[pp_fac.cov$p.adjusted < 0.05]

if (length(interaction_sig)>0){
 for (i in 1:length(interaction_sig)){
  group1 <- unlist(strsplit(interaction_sig[i],split = " vs "))[1]
  group2 <- unlist(strsplit(interaction_sig[i],split = " vs "))[2]
  group2 <- unlist(strsplit(group2,split = " : "))[1]
  
  result_list <- adonis_postanalysis(x=pairwise_df,
                      factors = filt_colon_metadata$Group,
                      covariate = filt_colon_metadata$Country, 
                      group1 = group1,
                      group2 = group2,
                      patients = filt_colon_metadata$Patient)
  print(result_list)
} 
}
```

##### PCoA

```{r, eval=FALSE}
colon_mpse <- as.MPSE(construct_phyloseq(filt_colon_asv_tab,filt_colon_taxa_tab,filt_colon_metadata))
colon_mpse %<>% mp_cal_dist(.abundance=Abundance, distmethod="robust.aitchison")
colon_mpse %<>% mp_cal_pcoa(.abundance=Abundance, distmethod="robust.aitchison")

# group-colored
a_plots <- pca_plots(colon_mpse,by="Group")

# country-colored
ac_plots <- pca_plots(colon_mpse,by="Country")
```

All in one

```{r, fig.width=10, fig.height=10, fig.fullwidth=TRUE, eval=FALSE}
a_12 <- ggarrange(a_plots[[1]],ac_plots[[1]],ncol=2)
a_13 <- ggarrange(a_plots[[2]],ac_plots[[2]],ncol=2)
a_23 <- ggarrange(a_plots[[3]],ac_plots[[3]],ncol=2)

a <- ggarrange(a_12, a_13, a_23, ncol=1)

# save the results
pca_plots_list[["Q2 Colon ASV"]] <- a

# see the results
a
```

##### PCoA custom

```{r}
beta_q2 <- pca_plot_custom(filt_colon_asv_tab,
                           filt_colon_taxa_tab,
                           filt_colon_metadata,
                       show_boxplots = TRUE,
                       variable = "Group", size=3, show_legend=TRUE)

# save the results
pca_plots_list[["Q2 Colon ASV custom"]] <- beta_q2

# see the results
beta_q2

```

#### Genus level

Aggregation, filtering

```{r}
genus_data <- aggregate_taxa(colon_asv_tab,
                             colon_taxa_tab,
                             taxonomic_level="Genus",
                             names=TRUE)

filt_data <- filtering_steps(genus_data[[1]],
                             genus_data[[2]],
                             colon_metadata,
                             seq_depth_threshold=10000)

filt_colon_genus_tab <- filt_data[[1]]
filt_colon_genus_taxa <- filt_data[[2]]
filt_colon_metadata <- filt_data[[3]]
```

##### PERMANOVA

```{r}
pairwise_df <- filt_colon_genus_tab %>% column_to_rownames("SeqID") %>% t()

# main effect
pp_main <- pairwise.adonis(pairwise_df,filt_colon_metadata$Group,
                           covariate = filt_colon_metadata$Country, 
                           patients = filt_colon_metadata$Patient,
                           sim.function = "robust.aitchison", p.adjust.m="BH")

# interaction
pp_int <- pairwise.adonis(pairwise_df,filt_colon_metadata$Group,
                          covariate = filt_colon_metadata$Country, 
                          interaction = TRUE, 
                          patients = filt_colon_metadata$Patient,
                          sim.function = "robust.aitchison", p.adjust.m="BH")

# tidy the results
pp_factor <- pp_main[[1]]
pp_cov <- pp_main[[2]]
pp_fac.cov <- pp_int[[3]]

cols <- c("pairs","Df","SumsOfSqs", "F.Model","R2","p.value", "p.adjusted", "sig")
colnames(pp_factor) <- cols; colnames(pp_cov) <- cols; colnames(pp_fac.cov) <- cols; 

# save raw results
pairwise_aitchison_raw[["Genus colon"]] <-rbind(pp_factor,pp_cov,pp_fac.cov)

# see the results
pp_factor
pp_cov
pp_fac.cov
```

Interaction check

```{r}
interaction_sig <- pp_fac.cov$pairs[pp_fac.cov$p.adjusted < 0.05]

if (length(interaction_sig)>0){
 for (i in 1:length(interaction_sig)){
  group1 <- unlist(strsplit(interaction_sig[i],split = " vs "))[1]
  group2 <- unlist(strsplit(interaction_sig[i],split = " vs "))[2]
  group2 <- unlist(strsplit(group2,split = " : "))[1]
  
  result_list <- adonis_postanalysis(x=pairwise_df,
                      factors = filt_colon_metadata$Group,
                      covariate = filt_colon_metadata$Country, 
                      group1 = group1,
                      group2 = group2,
                      patients = filt_colon_metadata$Patient)
  print(result_list)
} 
}

```

##### PCoA

```{r, eval=FALSE}
colon_mpse_genus <- as.MPSE(construct_phyloseq(filt_data[[1]],filt_data[[2]],filt_data[[3]]))
colon_mpse_genus %<>% mp_cal_dist(.abundance=Abundance, distmethod="robust.aitchison")
colon_mpse_genus %<>% mp_cal_pcoa(.abundance=Abundance, distmethod="robust.aitchison")

# group-colored
g_plots <- pca_plots(colon_mpse_genus,by="Group")

# country-colored
gc_plots <- pca_plots(colon_mpse_genus,by="Country")
```

All in one

```{r, fig.width=10, fig.height=10, fig.fullwidth=TRUE,eval=FALSE}
g_12 <- ggarrange(g_plots[[1]],gc_plots[[1]],ncol=2)
g_13 <- ggarrange(g_plots[[2]],gc_plots[[2]],ncol=2)
g_23 <- ggarrange(g_plots[[3]],gc_plots[[3]],ncol=2)

g <- ggarrange(g_12, g_13, g_23, ncol=1)

# save the results
pca_plots_list[["Q1 Colon Genus"]] <- g

# see the results
g
```

##### PCoA custom

```{r}
beta_q2_genus <- pca_plot_custom(filt_colon_genus_tab,
                                 filt_colon_genus_taxa,
                                 filt_colon_metadata,
                                 show_boxplots = TRUE,
                                 variable = "Group", size=2, 
                                 show_legend=TRUE)

# save the results
pca_plots_list[["Q2 Colon Genus custom"]] <- beta_q2_genus

# see the results
beta_q2_genus

```

### Univariate Analysis

#### ASV level

##### rPSC vs non-rPSC

###### linDA

```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
group <- c("non-rPSC","rPSC")

# prepare the data
linda_data <- binomial_prep(colon_asv_tab,
                            colon_taxa_tab,
                            colon_metadata,
                            group, usage="linDA")

filt_colon_uni_data <- linda_data[[1]]
filt_colon_uni_taxa <- linda_data[[2]]
filt_colon_uni_metadata <- linda_data[[3]]

# fit the model
linda.obj <- linda(filt_colon_uni_data, filt_colon_uni_metadata, formula = '~ Group * Country + (1|Patient)')

linda.output <- linda.obj$output
linda.output <- linda_renaming(linda.output, group)

# save the results
raw_linda_results[["colon"]] <- list()
linda_results[["colon"]] <- list()

group1 <- paste0(group[1], " vs ","Group",group[2])
group2 <- paste0(group[1], " , ",group[2], " - CZ vs NO") 
group3 <- paste0(group[1], " vs ","Group",group[2], ":CountryNO")

for (grp in c(group1,group2,group3)){
  raw_linda_results[["colon"]][[grp]] <- rawlinda.df(linda.output,
                                                     grp,
                                                     filt_colon_uni_data,
                                                     filt_colon_uni_taxa)
  linda_results[["colon"]][[grp]] <- linda.df(linda.output,
                                              grp,
                                              filt_colon_uni_data,
                                              filt_colon_uni_taxa)
}

# summary statistics
raw_linda_results <- binomial_statistics(filt_colon_uni_data, 
                                         filt_colon_uni_metadata,
                                         raw_linda_results, 
                                         group=group, 
                                         segment = "colon")


# for comparison
wb = createWorkbook()
new_name <- gsub("Group","",paste("linDa il A",group1))
addWorksheet(wb, sheetName = new_name)
linda_df <- rawlinda.df(linda.output,
                        group1,
                        filt_colon_uni_data,
                        filt_colon_uni_taxa)

writeData(wb, sheet = new_name, linda_df, rowNames=FALSE)

# see the results
raw_linda_results$colon[[group1]]
```

```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
# volcano plot
volcano_1 <- volcano_plot_linda(linda.output, group1,
                                taxa_table = filt_colon_uni_taxa) +
              ggtitle(paste(group,collapse=" vs "))

volcano_2  <- volcano_plot_linda(linda.output, group2, 
                                 taxa_table = filt_colon_uni_taxa) + 
              ggtitle("Country effect")

volcano_3 <- volcano_plot_linda(linda.output, group3, 
                                taxa_table = filt_colon_uni_taxa) + 
              ggtitle("Interaction")

volcano <- ggarrange(volcano_1,volcano_2,volcano_3, ncol=3)

# save the plot
volcano_all_linda[["colon"]] <- list(); i <- 1
volcano_all_linda[["colon"]][[i]] <- volcano; i <- i + 1

# see the plot
volcano
```

###### MaAsLin2

```{r, echo=FALSE,results='hide',message=FALSE,warning=FALSE}
fit_data = Maaslin2(
    input_data = filt_colon_uni_data, 
    input_metadata = filt_colon_uni_metadata, min_abundance = 0,
    min_prevalence = 0,min_variance = 0,
    output = paste0("results/Q2/",group1), max_significance = 0.05,
    fixed_effects = c('Group', 'Country'),random_effects = "Patient",correction = "BH")
```

```{r}
volcano1 <- volcano_plot_maaslin(fit_data,filt_colon_uni_taxa) + ggtitle(paste(group[1], "vs", group[2]))
volcano2 <- volcano_plot_maaslin(fit_data,filt_colon_uni_taxa,variable="Country") + ggtitle("Country effect")
volcano <- ggarrange(volcano1,volcano2, ncol=2)

volcano

# for comparison
new_name <- gsub("Group","",paste("Maas il A",group1))
addWorksheet(wb, sheetName = new_name)
maaslin_df <- raw_maaslin.df(fit_data,filt_colon_uni_data,filt_colon_uni_taxa)
writeData(wb, sheet = new_name, maaslin_df, rowNames=FALSE)

```

###### Group - Intersection

```{r}
intersection_results <- group_intersection(group, list_intersections, list_venns,
                                           linda.output, fit_data,
                                           raw_linda_results, 
                                           segment = "colon",
                                           level="ASV")

list_intersections <- intersection_results[[1]]
list_venns <- intersection_results[[2]]
venn <- intersection_results[[3]]

# show the results
venn
```

###### Country - Union

```{r, eval=FALSE}
list_country_union <- country_union(group,linda.output, fit_data,
                                    segment="colon",
                                    level="ASV")
```

###### Interaction effect

```{r}
list_interaction_significant <- country_interaction(group,linda.output, list_intersections,
                                          filt_colon_uni_data,
                                          filt_colon_uni_metadata,
                                          segment="colon",
                                          level="ASV")

# see the result
## significant interaction effect
list_interaction_significant[[1]]

## results for czech cohort
list_interaction_significant[[2]]

## results for norwegian cohort
list_interaction_significant[[3]]
```

Removing problematic taxa

```{r}
list_intersections <- removing_interaction_problems(group,
                                                    list_interaction_significant,
                                                    list_intersections,
                                                    segment="colon",
                                                    level="ASV")
```

##### rPSC vs healthy

###### linDA

```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
group <- c("healthy","rPSC")

# prepare the data
linda_data <- binomial_prep(colon_asv_tab,colon_taxa_tab,
                            colon_metadata,group, usage="linDA")

filt_colon_uni_data <- linda_data[[1]]
filt_colon_uni_taxa <- linda_data[[2]]
filt_colon_uni_metadata <- linda_data[[3]]

# fit the model
linda.obj <- linda(filt_colon_uni_data, filt_colon_uni_metadata, formula = '~ Group * Country + (1|Patient)')

linda.output <- linda.obj$output
linda.output <- linda_renaming(linda.output, group)

# save the results
group1 <- paste0(group[1], " vs ","Group",group[2])
group2 <- paste0(group[1], " , ",group[2], " - CZ vs NO") 
group3 <- paste0(group[1], " vs ","Group",group[2], ":CountryNO")

for (grp in c(group1,group2,group3)){
  raw_linda_results[["colon"]][[grp]] <- rawlinda.df(linda.output,grp,filt_colon_uni_data,filt_colon_uni_taxa)
  linda_results[["colon"]][[grp]] <- linda.df(linda.output,grp,filt_colon_uni_data,filt_colon_uni_taxa)
}

# summary statistics
raw_linda_results <- binomial_statistics(filt_colon_uni_data, filt_colon_uni_metadata, raw_linda_results, group=group, segment = "colon")

# for comparison
wb = createWorkbook()
new_name <- gsub("Group","",paste("linDa il A",group1))
addWorksheet(wb, sheetName = new_name)
linda_df <- rawlinda.df(linda.output,group1,filt_colon_uni_data,filt_colon_uni_taxa)
writeData(wb, sheet = new_name, linda_df, rowNames=FALSE)

# see the results
raw_linda_results$colon[[group1]]
```

```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
# volcano plot
volcano_1 <- volcano_plot_linda(linda.output, group1,
                                taxa_table = filt_colon_uni_taxa) +
              ggtitle(paste(group,collapse=" vs "))

volcano_2  <- volcano_plot_linda(linda.output, group2, 
                                 taxa_table = filt_colon_uni_taxa) + 
              ggtitle("Country effect")

volcano_3 <- volcano_plot_linda(linda.output, group3, 
                                taxa_table = filt_colon_uni_taxa) + 
              ggtitle("Interaction")

volcano <- ggarrange(volcano_1,volcano_2,volcano_3, ncol=3)

# save the plot
volcano_all_linda[["colon"]][[i]] <- volcano; i <- i + 1

# see the plot
volcano
```

###### MaAsLin2

```{r, echo=FALSE,results='hide',message=FALSE,warning=FALSE}
fit_data = Maaslin2(
    input_data = filt_colon_uni_data, 
    input_metadata = filt_colon_uni_metadata, min_abundance = 0,
    min_prevalence = 0,min_variance = 0,
    output = paste0("results/Q2/",group1), max_significance = 0.05,
    fixed_effects = c('Group', 'Country'),random_effects = "Patient",
    correction = "BH")

```

```{r, echo=FALSE}
volcano1 <- volcano_plot_maaslin(fit_data,filt_colon_uni_taxa) + ggtitle(paste(group[1], "vs", group[2]))
volcano2 <- volcano_plot_maaslin(fit_data,filt_colon_uni_taxa,variable="Country") + ggtitle("Country effect")
volcano <- ggarrange(volcano1,volcano2, ncol=2)
volcano
# for comparison
new_name <- gsub("Group","",paste("Maas il A",group1))
addWorksheet(wb, sheetName = new_name)
maaslin_df <- raw_maaslin.df(fit_data,filt_colon_uni_data,filt_colon_uni_taxa)
writeData(wb, sheet = new_name, maaslin_df, rowNames=FALSE)
```

###### Group - Intersection

```{r}
intersection_results <- group_intersection(group, list_intersections, list_venns,
                                           linda.output, fit_data,
                                           raw_linda_results, 
                                           segment = "colon",
                                           level="ASV")

list_intersections <- intersection_results[[1]]
list_venns <- intersection_results[[2]]
venn <- intersection_results[[3]]

# show the results
venn
```

###### Country - Union

```{r, eval=FALSE}
list_country_union <- country_union(group,linda.output, fit_data,
                                    segment="colon",
                                    level="ASV")
```

###### Interaction effect

```{r}
list_interaction_significant <- country_interaction(group,linda.output, list_intersections,
                                          filt_colon_uni_data,
                                          filt_colon_uni_metadata,
                                          segment="colon",
                                          level="ASV")

# see the result
## significant interaction effect
list_interaction_significant[[1]]

## results for czech cohort
list_interaction_significant[[2]]

## results for norwegian cohort
list_interaction_significant[[3]]
```

Removing problematic taxa

```{r}
list_intersections <- removing_interaction_problems(group,
                                                    list_interaction_significant,
                                                    list_intersections,
                                                    segment="colon",
                                                    level="ASV")
```

##### non-rPSC vs healthy

###### linDA

```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
group <- c("healthy","non-rPSC")

# prepare the data
linda_data <- binomial_prep(colon_asv_tab,colon_taxa_tab,
                            colon_metadata,group, usage="linDA")

filt_colon_uni_data <- linda_data[[1]]
filt_colon_uni_taxa <- linda_data[[2]]
filt_colon_uni_metadata <- linda_data[[3]]

# fit the model
linda.obj <- linda(filt_colon_uni_data, filt_colon_uni_metadata, formula = '~ Group * Country + (1|Patient)')

linda.output <- linda.obj$output
linda.output <- linda_renaming(linda.output, group)

# save the results
group1 <- paste0(group[1], " vs ","Group",group[2])
group2 <- paste0(group[1], " , ",group[2], " - CZ vs NO") 
group3 <- paste0(group[1], " vs ","Group",group[2], ":CountryNO")

for (grp in c(group1,group2,group3)){
  raw_linda_results[["colon"]][[grp]] <- rawlinda.df(linda.output,grp,filt_colon_uni_data,filt_colon_uni_taxa)
  linda_results[["colon"]][[grp]] <- linda.df(linda.output,grp,filt_colon_uni_data,filt_colon_uni_taxa)
}

# summary statistics
raw_linda_results <- binomial_statistics(filt_colon_uni_data, filt_colon_uni_metadata, raw_linda_results, group=group, segment = "colon")

# for comparison
wb = createWorkbook()
new_name <- gsub("Group","",paste("linDa il A",group1))
addWorksheet(wb, sheetName = new_name)
linda_df <- rawlinda.df(linda.output,group1,filt_colon_uni_data,filt_colon_uni_taxa)
writeData(wb, sheet = new_name, linda_df, rowNames=FALSE)

# see the results
raw_linda_results$colon[[group1]]
```

```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
# volcano plot
volcano_1 <- volcano_plot_linda(linda.output, group1,
                                taxa_table = filt_colon_uni_taxa) +
              ggtitle(paste(group,collapse=" vs "))

volcano_2  <- volcano_plot_linda(linda.output, group2, 
                                 taxa_table = filt_colon_uni_taxa) + 
              ggtitle("Country effect")

volcano_3 <- volcano_plot_linda(linda.output, group3, 
                                taxa_table = filt_colon_uni_taxa) + 
              ggtitle("Interaction")

volcano <- ggarrange(volcano_1,volcano_2,volcano_3, ncol=3)

# save the plot
volcano_all_linda[["colon"]][[i]] <- volcano; i <- i + 1

# see the plot
volcano
```

###### MaAsLin2

```{r, echo=FALSE,results='hide',message=FALSE,warning=FALSE}
fit_data = Maaslin2(
    input_data = filt_colon_uni_data, 
    input_metadata = filt_colon_uni_metadata, min_abundance = 0,
    min_prevalence = 0,min_variance = 0,
    output = paste0("results/Q2/",group1), max_significance = 0.05,
    fixed_effects = c('Group', 'Country'),random_effects = "Patient",
    correction = "BH")
```

```{r}
volcano1 <- volcano_plot_maaslin(fit_data,filt_colon_uni_taxa) + ggtitle(paste(group[1], "vs", group[2]))
volcano2 <- volcano_plot_maaslin(fit_data,filt_colon_uni_taxa,variable="Country") + ggtitle("Country effect")
volcano <- ggarrange(volcano1,volcano2, ncol=2)
volcano
# for comparison
new_name <- gsub("Group","",paste("Maas il A",group1))
addWorksheet(wb, sheetName = new_name)
maaslin_df <- raw_maaslin.df(fit_data,filt_colon_uni_data,filt_colon_uni_taxa)
writeData(wb, sheet = new_name, maaslin_df, rowNames=FALSE)
```

###### Group - Intersection

```{r}
intersection_results <- group_intersection(group, list_intersections, list_venns,
                                           linda.output, fit_data,
                                           raw_linda_results, 
                                           segment = "colon",
                                           level="ASV")

list_intersections <- intersection_results[[1]]
list_venns <- intersection_results[[2]]
venn <- intersection_results[[3]]

# show the results
venn
```

###### Country - Union

```{r, eval=FALSE}
list_country_union <- country_union(group,linda.output, fit_data,
                                    segment="colon",
                                    level="ASV")
```

###### Interaction effect

```{r}
list_interaction_significant <- country_interaction(group,linda.output, list_intersections,
                                          filt_colon_uni_data,
                                          filt_colon_uni_metadata,
                                          segment="colon",
                                          level="ASV")

# see the result
## significant interaction effect
list_interaction_significant[[1]]

## results for czech cohort
list_interaction_significant[[2]]

## results for norwegian cohort
list_interaction_significant[[3]]
```

Removing problematic taxa

```{r}
list_intersections <- removing_interaction_problems(group,
                                                    list_interaction_significant,
                                                    list_intersections,
                                                    segment="colon",
                                                    level="ASV")
```

##### Heatmap

Heatmap visualizing the linDA's logFoldChange for taxa with p \< 0.1.

```{r, fig.width=15, fig.height=20, fig.fullwidth=TRUE}
list_heatmap <- list_intersections[grep("colon ASV",names(list_intersections),value=TRUE)]
heatmap_asv_linda <- heatmap_linda(list_heatmap,colon_taxa_tab)
heatmap_asv_linda
```

##### Dot heatmap

```{r, fig.width=5, fig.height=25, fig.fullwidth=TRUE}
dotheatmap_linda <- dot_heatmap_linda(list_heatmap,colon_taxa_tab)
dotheatmap_linda
```

##### rPSC effect

**rPSC vs Healthy and non-rPSC vs Healthy LEFT JOIN**

![imagename](pre_post.png)

```{r}
A <- list_intersections$`colon ASV healthy vs rPSC`
B <- list_intersections$`colon ASV healthy vs non-rPSC`

df <- A[!(A$ASV %in% B$ASV),]

rpsc_effect[["colon ASV"]] <- df

# see the results
rpsc_effect[["colon ASV"]]
```

#### Genus level

Aggregate taxa

```{r}
genus_data <- aggregate_taxa(colon_asv_tab,colon_taxa_tab,taxonomic_level = "Genus")

colon_genus_tab <- genus_data[[1]]
colon_genus_taxa_tab <- genus_data[[2]]

colon_genus_asv_taxa_tab <- create_asv_taxa_table(colon_genus_tab,colon_genus_taxa_tab)
```

##### rPSC vs non-rPSC

###### linDA

```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
group <- c("non-rPSC","rPSC")

# prepare the data
linda_data <- binomial_prep(colon_genus_tab,colon_genus_taxa_tab,
                            colon_metadata,group, usage="linDA")

filt_colon_uni_data <- linda_data[[1]]
filt_colon_uni_taxa <- linda_data[[2]]
filt_colon_uni_metadata <- linda_data[[3]]

# fit the model
linda.obj <- linda(filt_colon_uni_data, filt_colon_uni_metadata, formula = '~ Group * Country + (1|Patient)')

linda.output <- linda.obj$output
linda.output <- linda_renaming(linda.output, group)

# save the results
group1 <- paste0(group[1], " vs ","Group",group[2])
group2 <- paste0(group[1], " , ",group[2], " - CZ vs NO") 
group3 <- paste0(group[1], " vs ","Group",group[2], ":CountryNO")

raw_linda_results_genus[["colon"]] <- list()
linda_results_genus[["colon"]] <- list()

for (grp in c(group1,group2,group3)){
  raw_linda_results_genus[["colon"]][[grp]] <- rawlinda.df(linda.output,grp,filt_colon_uni_data,filt_colon_uni_taxa)
  linda_results_genus[["colon"]][[grp]] <- linda.df(linda.output,grp,filt_colon_uni_data,filt_colon_uni_taxa)
}

# summary statistics
raw_linda_results_genus <- binomial_statistics(filt_colon_uni_data,                     
                                            group=group,
                                            filt_colon_uni_metadata,
                                            raw_linda_results_genus,
                                            segment = "colon")

# for comparison
new_name <- gsub("Group","",paste("linDA il G",group1))
addWorksheet(wb, sheetName = new_name)
linda_df <- rawlinda.df(linda.output,group1,filt_colon_uni_data,filt_colon_uni_taxa)
writeData(wb, sheet = new_name, linda_df, rowNames=FALSE)

```

```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
# volcano plot
volcano_1 <- volcano_plot_linda(linda.output, group1, 
                                taxa_table = filt_colon_uni_taxa) + 
            ggtitle(paste(group,collapse=" vs "))

volcano_2  <- volcano_plot_linda(linda.output, group2, 
                                taxa_table = filt_colon_uni_taxa) + 
            ggtitle("Country effect") 

volcano_3 <- volcano_plot_linda(linda.output, group3, 
                                taxa_table = filt_colon_uni_taxa) +
            ggtitle("Interaction")

volcano <- ggarrange(volcano_1,volcano_2,volcano_3, ncol=3)

# save the plot
volcano_all_linda_genus[["colon"]][[i]] <- volcano; i <- i + 1

# see the plot
volcano
```

###### MaAsLin2

```{r, echo=FALSE,results='hide',message=FALSE,warning=FALSE}
fit_data = Maaslin2(
    input_data = filt_colon_uni_data, 
    input_metadata = filt_colon_uni_metadata, min_abundance = 0,
    min_prevalence = 0,min_variance = 0,
    output = paste0("results/Q2/G",group1), max_significance = 0.05,
    fixed_effects = c('Group', 'Country'),correction = "BH",
    random_effects = "Patient")
```

```{r}
volcano1 <- volcano_plot_maaslin(fit_data,filt_colon_uni_taxa) + ggtitle(paste(group[1], "vs", group[2]))
volcano2 <- volcano_plot_maaslin(fit_data,filt_colon_uni_taxa,variable="Country") + ggtitle("Country effect")
volcano <- ggarrange(volcano1,volcano2, ncol=2)
volcano
# for comparison
new_name <- gsub("Group","",paste("Maas c G",group1))
addWorksheet(wb, sheetName = new_name)
maaslin_df <- raw_maaslin.df(fit_data,filt_colon_uni_data,filt_colon_uni_taxa)
writeData(wb, sheet = new_name, maaslin_df, rowNames=FALSE)
```

###### Group - Intersection

```{r}
intersection_results <- group_intersection(group, list_intersections, list_venns,
                                           linda.output, fit_data,
                                           raw_linda_results_genus, 
                                           segment = "colon",
                                           level="Genus")

list_intersections <- intersection_results[[1]]
list_venns <- intersection_results[[2]]
venn <- intersection_results[[3]]

# show the results
venn
```

###### Country - Union

```{r, eval=FALSE}
list_country_union <- country_union(group,linda.output, fit_data,
                                    segment="colon",
                                    level="Genus")
```

###### Interaction effect

```{r}
list_interaction_significant <- country_interaction(group,linda.output, list_intersections,
                                          filt_colon_uni_data,
                                          filt_colon_uni_metadata,
                                          segment="colon",
                                          level="Genus")

# see the result
## significant interaction effect
list_interaction_significant[[1]]

## results for czech cohort
list_interaction_significant[[2]]

## results for norwegian cohort
list_interaction_significant[[3]]
```

Removing problematic taxa

```{r}
list_intersections <- removing_interaction_problems(group,
                                                    list_interaction_significant,
                                                    list_intersections,
                                                    segment="colon",
                                                    level="Genus")
```

##### rPSC vs healthy

###### linDA

```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
# group <- c("rPSC","healthy")
group <- c("healthy","rPSC")

# prepare the data
linda_data <- binomial_prep(colon_genus_tab,colon_genus_taxa_tab,
                            colon_metadata,group, usage="linDA")

filt_colon_uni_data <- linda_data[[1]]
filt_colon_uni_taxa <- linda_data[[2]]
filt_colon_uni_metadata <- linda_data[[3]]

# fit the model
linda.obj <- linda(filt_colon_uni_data, filt_colon_uni_metadata, formula = '~ Group * Country + (1|Patient)')

linda.output <- linda.obj$output
linda.output <- linda_renaming(linda.output, group)

# save the results
group1 <- paste0(group[1], " vs ","Group",group[2])
group2 <- paste0(group[1], " , ",group[2], " - CZ vs NO") 
group3 <- paste0(group[1], " vs ","Group",group[2], ":CountryNO")

raw_linda_results_genus[["colon"]] <- list()
linda_results_genus[["colon"]] <- list()

for (grp in c(group1,group2,group3)){
  raw_linda_results_genus[["colon"]][[grp]] <- rawlinda.df(linda.output,grp,filt_colon_uni_data,filt_colon_uni_taxa)
  linda_results_genus[["colon"]][[grp]] <- linda.df(linda.output,grp,filt_colon_uni_data,filt_colon_uni_taxa)
}

# summary statistics
raw_linda_results_genus <- binomial_statistics(filt_colon_uni_data,                     
                                            group=group,
                                            filt_colon_uni_metadata,
                                            raw_linda_results_genus,
                                            segment = "colon")

# for comparison
new_name <- gsub("Group","",paste("linDA il G",group1))
addWorksheet(wb, sheetName = new_name)
linda_df <- rawlinda.df(linda.output,group1,filt_colon_uni_data,filt_colon_uni_taxa)
writeData(wb, sheet = new_name, linda_df, rowNames=FALSE)

```


```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
# volcano plot
volcano_1 <- volcano_plot_linda(linda.output, group1, 
                                taxa_table = filt_colon_uni_taxa) + 
            ggtitle(paste(group,collapse=" vs "))

volcano_2  <- volcano_plot_linda(linda.output, group2, 
                                taxa_table = filt_colon_uni_taxa) + 
            ggtitle("Country effect") 

volcano_3 <- volcano_plot_linda(linda.output, group3, 
                                taxa_table = filt_colon_uni_taxa) +
            ggtitle("Interaction")

volcano <- ggarrange(volcano_1,volcano_2,volcano_3, ncol=3)

# save the plot
volcano_all_linda_genus[["colon"]][[i]] <- volcano; i <- i + 1

# see the plot
volcano
```

###### MaAsLin2

```{r, echo=FALSE,results='hide',message=FALSE,warning=FALSE}
fit_data = Maaslin2(
    input_data = filt_colon_uni_data, 
    input_metadata = filt_colon_uni_metadata, min_abundance = 0,
    min_prevalence = 0,min_variance = 0,
    output = paste0("results/Q2/G",group1), max_significance = 0.05,
    fixed_effects = c('Group', 'Country'),random_effects = "Patient",
    correction = "BH")

```

```{r}
volcano1 <- volcano_plot_maaslin(fit_data,filt_colon_uni_taxa) + ggtitle(paste(group[1], "vs", group[2]))
volcano2 <- volcano_plot_maaslin(fit_data,filt_colon_uni_taxa,variable="Country") + ggtitle("Country effect")
volcano <- ggarrange(volcano1,volcano2, ncol=2)
volcano
# for comparison
new_name <- gsub("Group","",paste("Maas c G",group1))
addWorksheet(wb, sheetName = new_name)
maaslin_df <- raw_maaslin.df(fit_data,filt_colon_uni_data,filt_colon_uni_taxa)
writeData(wb, sheet = new_name, maaslin_df, rowNames=FALSE)
```

###### Group - Intersection

```{r}
intersection_results <- group_intersection(group, list_intersections, list_venns,
                                           linda.output, fit_data,
                                           raw_linda_results_genus, 
                                           segment = "colon",
                                           level="Genus")

list_intersections <- intersection_results[[1]]
list_venns <- intersection_results[[2]]
venn <- intersection_results[[3]]

# show the results
venn
```

###### Country - Union

```{r, eval=FALSE}
list_country_union <- country_union(group,linda.output, fit_data,
                                    segment="colon",
                                    level="Genus")
```

###### Interaction effect

```{r}
list_interaction_significant <- country_interaction(group,linda.output, list_intersections,
                                          filt_colon_uni_data,
                                          filt_colon_uni_metadata,
                                          segment="colon",
                                          level="Genus")

# see the result
## significant interaction effect
list_interaction_significant[[1]]

## results for czech cohort
list_interaction_significant[[2]]

## results for norwegian cohort
list_interaction_significant[[3]]
```

Removing problematic taxa

```{r}
list_intersections <- removing_interaction_problems(group,
                                                    list_interaction_significant,
                                                    list_intersections,
                                                    segment="colon",
                                                    level="Genus")
```

##### non-rPSC vs healthy

###### linDA

```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
#group <- c("non-rPSC","healthy")
group <- c("healthy","non-rPSC")

# prepare the data
linda_data <- binomial_prep(colon_genus_tab,colon_genus_taxa_tab,
                            colon_metadata,group, usage="linDA")

filt_colon_uni_data <- linda_data[[1]]
filt_colon_uni_taxa <- linda_data[[2]]
filt_colon_uni_metadata <- linda_data[[3]]

# fit the model
linda.obj <- linda(filt_colon_uni_data, filt_colon_uni_metadata, formula = '~ Group * Country + (1|Patient)')

linda.output <- linda.obj$output
linda.output <- linda_renaming(linda.output, group)

# save the results
group1 <- paste0(group[1], " vs ","Group",group[2])
group2 <- paste0(group[1], " , ",group[2], " - CZ vs NO") 
group3 <- paste0(group[1], " vs ","Group",group[2], ":CountryNO")

raw_linda_results_genus[["colon"]] <- list()
linda_results_genus[["colon"]] <- list()

for (grp in c(group1,group2,group3)){
  raw_linda_results_genus[["colon"]][[grp]] <- rawlinda.df(linda.output,grp,filt_colon_uni_data,filt_colon_uni_taxa)
  linda_results_genus[["colon"]][[grp]] <- linda.df(linda.output,grp,filt_colon_uni_data,filt_colon_uni_taxa)
}

# summary statistics
raw_linda_results_genus <- binomial_statistics(filt_colon_uni_data,                     
                                            group=group,
                                            filt_colon_uni_metadata,
                                            raw_linda_results_genus,
                                            segment = "colon")

# for comparison
new_name <- gsub("Group","",paste("linDA il G",group1))
addWorksheet(wb, sheetName = new_name)
linda_df <- rawlinda.df(linda.output,group1,filt_colon_uni_data,filt_colon_uni_taxa)
writeData(wb, sheet = new_name, linda_df, rowNames=FALSE)

```


```{r, fig.width=15, fig.height=5, fig.fullwidth=TRUE}
# volcano plot
volcano_1 <- volcano_plot_linda(linda.output, group1, 
                                taxa_table = filt_colon_uni_taxa) + 
            ggtitle(paste(group,collapse=" vs "))

volcano_2  <- volcano_plot_linda(linda.output, group2, 
                                taxa_table = filt_colon_uni_taxa) + 
            ggtitle("Country effect") 

volcano_3 <- volcano_plot_linda(linda.output, group3, 
                                taxa_table = filt_colon_uni_taxa) +
            ggtitle("Interaction")

volcano <- ggarrange(volcano_1,volcano_2,volcano_3, ncol=3)

# save the plot
volcano_all_linda_genus[["colon"]][[i]] <- volcano; i <- i + 1

# see the plot
volcano
```

###### MaAsLin2

```{r, echo=FALSE,results='hide',message=FALSE,warning=FALSE}
fit_data = Maaslin2(
    input_data = filt_colon_uni_data, 
    input_metadata = filt_colon_uni_metadata, min_abundance = 0,
    min_prevalence = 0,min_variance = 0,
    output = paste0("results/Q2/G",group1), max_significance = 0.05,
    fixed_effects = c('Group', 'Country'),random_effects = "Patient",
    correction = "BH")
```

```{r}
volcano1 <- volcano_plot_maaslin(fit_data,filt_colon_uni_taxa) + ggtitle(paste(group[1], "vs", group[2]))
volcano2 <- volcano_plot_maaslin(fit_data,filt_colon_uni_taxa,variable="Country") + ggtitle("Country effect")
volcano <- ggarrange(volcano1,volcano2, ncol=2)
volcano

# for comparison
new_name <- gsub("Group","",paste("Maas c G",group1))
addWorksheet(wb, sheetName = new_name)
maaslin_df <- raw_maaslin.df(fit_data,filt_colon_uni_data,filt_colon_uni_taxa)
writeData(wb, sheet = new_name, maaslin_df, rowNames=FALSE)
```

###### Group - Intersection

```{r}
intersection_results <- group_intersection(group, list_intersections, list_venns,
                                           linda.output, fit_data,
                                           raw_linda_results_genus, 
                                           segment = "colon",
                                           level="Genus")

list_intersections <- intersection_results[[1]]
list_venns <- intersection_results[[2]]
venn <- intersection_results[[3]]

# show the results
venn
```

###### Country - Union

```{r, eval=FALSE}
list_country_union <- country_union(group,linda.output, fit_data,
                                    segment="colon",
                                    level="Genus")
```

###### Interaction effect

```{r}
list_interaction_significant <- country_interaction(group,linda.output, list_intersections,
                                          filt_colon_uni_data,
                                          filt_colon_uni_metadata,
                                          segment="colon",
                                          level="Genus")

# see the result
## significant interaction effect
list_interaction_significant[[1]]

## results for czech cohort
list_interaction_significant[[2]]

## results for norwegian cohort
list_interaction_significant[[3]]
```

Removing problematic taxa

```{r}
list_intersections <- removing_interaction_problems(group,
                                                    list_interaction_significant,
                                                    list_intersections,
                                                    segment="colon",
                                                    level="Genus")
```

##### Visualizations

Heatmap visualizing the linDA's logFoldChange for taxa with p \< 0.1.

```{r, fig.width=10, fig.height=10, fig.fullwidth=TRUE}
list_heatmap <- list_intersections[grep("colon Genus",names(list_intersections),value=TRUE)]
heatmap_asv_linda <- heatmap_linda(list_heatmap,colon_taxa_tab)
heatmap_asv_linda
```

Dot heatmap

```{r, fig.width=5, fig.height=15, fig.fullwidth=TRUE}
dotheatmap_linda <- dot_heatmap_linda(list_heatmap,colon_taxa_tab)
dotheatmap_linda
```

##### rPSC effect

**rPSC vs Healthy and non-rPSC vs Healthy LEFT JOIN**

![imagename](pre_post.png)

```{r}
A <- list_intersections$`colon Genus healthy vs rPSC`
B <- list_intersections$`colon Genus healthy vs non-rPSC`

df <- A[!(A$ASV %in% B$ASV),]

rpsc_effect[["colon Genus"]] <- df

# see the results
rpsc_effect[["colon Genus"]]
```

### ElasticNet - merged cohorts

#### ASV level

##### rPSC vs non-rPSC

```{r}
group <- c("rPSC","non-rPSC")

model_name <- paste(group[1],"vs",group[2],"ASV colon")
  
# prepare the data
filt_colon_uni_data <- binomial_prep(colon_asv_tab,colon_taxa_tab,
                                     colon_metadata,group, usage="glmnet",
                                     patient=TRUE)

# fit the model
enet_model <- glmnet_binomial(filt_colon_uni_data,sample_method = "atypboot",
                                 outcome="Group", N=500,
                                 clust_var="Patient",
                              reuse=TRUE,
                              file=model_name,
                              Q="Q2")

# save the results
models_summ[[paste(group[1],"vs",group[2],"ASV colon")]] <- enet_model$model_summary
models_cm[[paste(group[1],"vs",group[2],"ASV colon")]] <- enet_model$conf_matrices$original
roc_c <- roc_curve(enet_model, group)
roc_cs[[paste(group[1],"vs",group[2],"ASV colon")]] <- enet_model$kfold_rocobjs

# see the results
enet_model$model_summary %>% t()
enet_model$conf_matrices
enet_model$plot

roc_c
```

##### rPSC vs healthy

```{r}
group <- c("rPSC","healthy")
model_name <- paste(group[1],"vs",group[2],"ASV colon")

# prepare the data
filt_colon_uni_data <- binomial_prep(colon_asv_tab,colon_taxa_tab,
                                            colon_metadata,group, usage="glmnet",
                                     patient=TRUE)

# fit the model
enet_model <- glmnet_binomial(filt_colon_uni_data,sample_method = "atypboot",
                                 outcome="Group", N=500,
                                 clust_var="Patient",
                              reuse=TRUE,
                              file=model_name,
                              Q="Q2")

# save the results
models_summ[[paste(group[1],"vs",group[2],"ASV colon")]] <- enet_model$model_summary
models_cm[[paste(group[1],"vs",group[2],"ASV colon")]] <- enet_model$conf_matrices$original
roc_c <- roc_curve(enet_model, group)
roc_cs[[paste(group[1],"vs",group[2],"ASV colon")]] <- enet_model$kfold_rocobjs

# see the results
enet_model$model_summary %>% t()
enet_model$conf_matrices
enet_model$plot

roc_c
```

##### non-rPSC vs healthy

```{r}
group <- c("non-rPSC","healthy")
model_name <- paste(group[1],"vs",group[2],"ASV colon")

# prepare the data
filt_colon_uni_data <- binomial_prep(colon_asv_tab,colon_taxa_tab,
                                            colon_metadata,group, usage="glmnet",
                                     patient=TRUE)

# fit the model
enet_model <- glmnet_binomial(filt_colon_uni_data,sample_method = "atypboot",
                                 outcome="Group", N=500,
                                 clust_var="Patient",
                              reuse=TRUE,
                              file=model_name,
                              Q="Q2")

# save the results
models_summ[[paste(group[1],"vs",group[2],"ASV colon")]] <- enet_model$model_summary
models_cm[[paste(group[1],"vs",group[2],"ASV colon")]] <- enet_model$conf_matrices$original
roc_c <- roc_curve(enet_model, group)
roc_cs[[paste(group[1],"vs",group[2],"ASV colon")]] <- enet_model$kfold_rocobjs

# see the results
enet_model$model_summary %>% t()
enet_model$conf_matrices
enet_model$plot

roc_c
```

##### rPSC effect

```{r}
model_name <- paste("rPSC effect","ASV colon")
# prepare the data
filt_ileum_uni_data <- binomial_prep_psc_effect(colon_asv_tab,colon_taxa_tab, 
                                     colon_metadata,rpsc_effect$`colon ASV`,
                                     patient = TRUE,
                                     usage="glmnet")

# fit the model
enet_model <- glmnet_binomial(filt_ileum_uni_data,sample_method = "atypboot",
                                 outcome="Group", N=500, clust_var = "Patient",
                              reuse=TRUE,
                              file=model_name,
                              Q="Q2")

# save the results
models_summ[[paste("rPSC effect","ASV colon")]] <- enet_model$model_summary
models_cm[[paste("rPSC effect","ASV colon")]] <- enet_model$conf_matrices$original
roc_c <- roc_curve(enet_model, group)
roc_cs[[paste("rPSC effect","ASV colon")]] <- enet_model$kfold_rocobjs

# see the results
enet_model$model_summary %>% t()
enet_model$conf_matrices
enet_model$plot

roc_c
```

#### Genus level

Aggregate taxa

```{r}
genus_data <- aggregate_taxa(colon_asv_tab,colon_taxa_tab,taxonomic_level = "Genus")

colon_genus_tab <- genus_data[[1]]
colon_genus_taxa_tab <- genus_data[[2]]
```

##### rPSC vs non-rPSC

```{r}
group <- c("rPSC","non-rPSC")
model_name <- paste(group[1],"vs",group[2],"Genus colon")

# prepare the data
filt_colon_uni_data <- binomial_prep(colon_genus_tab,colon_genus_taxa_tab,
                                            colon_metadata,group, usage="glmnet",
                                     patient=TRUE)

# fit the model
enet_model <- glmnet_binomial(filt_colon_uni_data,sample_method = "atypboot",
                                 outcome="Group", N=500,
                                 clust_var="Patient",
                              reuse=TRUE,
                              file=model_name,
                              Q="Q2")

# save the results
models_summ[[paste(group[1],"vs",group[2],"Genus colon")]] <- enet_model$model_summary
models_cm[[paste(group[1],"vs",group[2],"Genus colon")]] <- enet_model$conf_matrices$original
roc_c <- roc_curve(enet_model, group)
roc_cs[[paste(group[1],"vs",group[2],"Genus colon")]] <- enet_model$kfold_rocobjs

# see the results
enet_model$model_summary %>% t()
enet_model$conf_matrices
enet_model$plot

roc_c
```

##### rPSC vs healthy

```{r}
group <- c("rPSC","healthy")
model_name <- paste(group[1],"vs",group[2],"Genus colon")

# prepare the data
filt_colon_uni_data <- binomial_prep(colon_genus_tab,colon_genus_taxa_tab,
                                            colon_metadata,group, usage="glmnet",
                                     patient=TRUE)

# fit the model
enet_model <- glmnet_binomial(filt_colon_uni_data,sample_method = "atypboot",
                                 outcome="Group", N=500, 
                                 clust_var="Patient",
                              reuse=TRUE,
                              file=model_name,
                              Q="Q2")

# save the results
models_summ[[paste(group[1],"vs",group[2],"Genus colon")]] <- enet_model$model_summary
models_cm[[paste(group[1],"vs",group[2],"Genus colon")]] <- enet_model$conf_matrices$original
roc_c <- roc_curve(enet_model, group)
roc_cs[[paste(group[1],"vs",group[2],"Genus colon")]] <- enet_model$kfold_rocobjs

# see the results
enet_model$model_summary %>% t()
enet_model$conf_matrices
enet_model$plot

roc_c
```

##### non-rPSC vs healthy

```{r}
group <- c("non-rPSC","healthy")
model_name <- paste(group[1],"vs",group[2],"Genus colon")

# prepare the data
filt_colon_uni_data <- binomial_prep(colon_genus_tab,colon_genus_taxa_tab,
                                            colon_metadata,group, usage="glmnet",
                                     patient=TRUE)

# fit the model
enet_model <- glmnet_binomial(filt_colon_uni_data,sample_method = "atypboot",
                                 outcome="Group", N=500, 
                                 clust_var="Patient",
                              reuse=TRUE,
                              file=model_name,
                              Q="Q2")

# save the results
models_summ[[paste(group[1],"vs",group[2],"Genus colon")]] <- enet_model$model_summary
models_cm[[paste(group[1],"vs",group[2],"Genus colon")]] <- enet_model$conf_matrices$original
roc_c <- roc_curve(enet_model, group)
roc_cs[[paste(group[1],"vs",group[2],"Genus colon")]] <- enet_model$kfold_rocobjs

# see the results
enet_model$model_summary %>% t()
enet_model$conf_matrices
enet_model$plot

roc_c
```

##### rPSC effect

```{r}
model_name <- paste("rPSC effect","Genus colon")
# prepare the data
filt_ileum_uni_data <- binomial_prep_psc_effect(colon_genus_tab,colon_genus_taxa_tab, 
                                     colon_metadata,rpsc_effect$`colon Genus`,
                                     patient = TRUE,
                                     usage="glmnet")

# fit the model
enet_model <- glmnet_binomial(filt_ileum_uni_data,sample_method = "atypboot",
                                 outcome="Group", N=500,
                              clust_var = "Patient",
                              reuse=TRUE,
                              file=model_name,
                              Q="Q2")

# save the results
models_summ[[paste("rPSC effect","Genus colon")]] <- enet_model$model_summary
models_cm[[paste("rPSC effect","Genus colon")]] <- enet_model$conf_matrices$original
roc_c <- roc_curve(enet_model, group)
roc_cs[[paste("rPSC effect","Genus colon")]] <- enet_model$kfold_rocobjs

# see the results
enet_model$model_summary %>% t()
enet_model$conf_matrices
enet_model$plot

roc_c
```

### Results overview

#### Alpha diversity

```{r}
pc_observed$colon
pc_shannon$colon
pc_simpson$colon
pc_pielou$colon

write.xlsx(x = list(pc_observed$colon,
                    pc_shannon$colon,
                    pc_simpson$colon,
                    pc_pielou$colon), 
           file = "results/Q2/alpha_diversity_results_colon.xlsx")
```


```{r, fig.width=10, fig.height=4, fig.fullwidth=TRUE}
alpha_div_plots[["Colon Country"]]
```

```{r, fig.width=10, fig.height=4, fig.fullwidth=TRUE}
alpha_div_plots[["Colon Custom"]]
```

#### Beta diversity

**Raw results at ASV and Genus level**

```{r}
pairwise_aitchison_raw$`ASV colon`
pairwise_aitchison_raw$`Genus colon`
```

**PCA at ASV level**

```{r, fig.width=10, fig.height=10, fig.fullwidth=TRUE}
pca_plots_list$`Q1 Colon ASV`
```

**PCA at Genus level**

```{r, fig.width=10, fig.height=10, fig.fullwidth=TRUE}
pca_plots_list$`Q2 colon Genus`
```

#### Univariate analysis

**Number of significant taxa**

```{r}
cbind(as.data.frame(
  lapply(list_intersections,nrow)),
  as.data.frame(
  lapply(rpsc_effect,nrow))) %>%
  t() %>% 
  `colnames<-`("Count") #%>% 
  #`rownames<-`(c(names(list_intersections),"PSC effect ASV","PSC effect Genus"))
```

```{r}
# save the results
write.xlsx(x = list_intersections[grep("colon",names(list_intersections),value=TRUE)],
           file = "results/Q2/DAA_colon.xlsx")

write.xlsx(x = rpsc_effect[grep("colon",names(rpsc_effect),value=TRUE)], 
           file = "results/Q2/rPSC_effect_colon.xlsx")

saveWorkbook(wb,"results/Q2/DAA_linDA_Maaslin_colon.xlsx", overwrite = TRUE)
```

##### How significant ASVs and genera overlap

1. Is every single significant ASV also in a significant genus? If not, this ASV points to a strain-specific biomarker. 

2. Does every significant genus have at least one significant ASV?
It should be the case, but if not, visualize this genus and all ASVs that belong to it, try to find out why it is not the case.

```{r}
univar_list <- univariate_statistics(list_intersections,
                                     rpsc_effect,
                                     colon_genus_asv_taxa_tab, segment="colon")

univar_df <- univar_list[[1]]
wb <- univar_list[[2]]

# save the results
saveWorkbook(wb,"results/Q2/DAA_final_colon.xlsx", overwrite = TRUE)

# see the results
univar_df
```


##### Terminal ileum and colon overlap

###### ASV level

rPSC effect

```{r}
psc_effect_TI <- read.xlsx("results/Q2/rPSC_effect_terminal_ileum.xlsx",
                           sheet = "Ileum ASV")
psc_effect_colon <- read.xlsx("results/Q2/rPSC_effect_colon.xlsx",
                              sheet = "colon ASV")

intersection <- intersect(psc_effect_TI$ASV,psc_effect_colon$ASV)
intersection_psc <- psc_effect_TI[psc_effect_TI$ASV %in% intersection,]
intersection_psc
```

rPSC vs non-rPSC

```{r}
r_nonr_TI <- read.xlsx("results/Q2/DAA_final_terminal_ileum.xlsx",
                       sheet = " non-rPSC vs rPSC ASVs")
r_nonr_colon <- read.xlsx("results/Q2/DAA_final_colon.xlsx",
                          sheet = " non-rPSC vs rPSC ASVs")

intersection <- intersect(r_nonr_TI$ASV,r_nonr_colon$ASV)
intersection_r_nonr <- r_nonr_TI[r_nonr_TI$ASV %in% intersection,]
intersection_r_nonr
```

rPSC vs healthy

```{r}
r_H_TI <- read.xlsx("results/Q2/DAA_final_terminal_ileum.xlsx",
                    sheet = " H vs rPSC ASVs")
r_H_colon <- read.xlsx("results/Q2/DAA_final_colon.xlsx",
                       sheet = " H vs rPSC ASVs")

intersection <- intersect(r_H_TI$ASV,r_H_colon$ASV)
intersection_r_H <- r_H_TI[r_H_TI$ASV %in% intersection,]
intersection_r_H
```

non-rPSC vs healthy

```{r}
non_H_TI <- read.xlsx("results/Q2/DAA_final_terminal_ileum.xlsx",
                      sheet = " H vs non-rPSC ASVs")
non_H_colon <- read.xlsx("results/Q2/DAA_final_colon.xlsx",sheet = " H vs non-rPSC ASVs")

intersection <- intersect(non_H_TI$ASV,non_H_colon$ASV)
intersection_nonr_H <- non_H_TI[non_H_TI$ASV %in% intersection,]
intersection_nonr_H
```

Statistics

```{r}
df <- data.frame(`Number of ASVs`= c(nrow(intersection_r_nonr),
                               nrow(intersection_r_H),
                               nrow(intersection_nonr_H),
                               nrow(intersection_psc)),
           `Unique genera`=c(length(unique(intersection_r_nonr$Taxonomy)),
                             length(unique(intersection_r_H$Taxonomy)),
                             length(unique(intersection_nonr_H$Taxonomy)),
                             length(unique(intersection_psc$Taxonomy)))) 

rownames(df) <- c("rPSC vs non-rPSC", "rPSC vs healthy", "non-rPSC vs healthy","rPSC effect")

df
```

###### Genus level

PSC effect
```{r}
psc_effect_TI <- read.xlsx("results/Q2/rPSC_effect_terminal_ileum.xlsx",sheet = "Ileum Genus")
psc_effect_colon <- read.xlsx("results/Q2/rPSC_effect_colon.xlsx",sheet = "colon Genus")

intersection <- intersect(psc_effect_TI$ASV,psc_effect_colon$ASV)
intersection_psc <- psc_effect_TI[psc_effect_TI$ASV %in% intersection,]
intersection_psc
```

rPSC vs non-rPSC

```{r}
r_nonr_TI <- read.xlsx("results/Q2/DAA_final_terminal_ileum.xlsx",
                       sheet = " non-rPSC vs rPSC Genera")
r_nonr_colon <- read.xlsx("results/Q2/DAA_final_colon.xlsx",
                          sheet = " non-rPSC vs rPSC Genera")
intersection <- intersect(r_nonr_TI$ASV,r_nonr_colon$ASV)
intersection_r_nonr <- r_nonr_TI[r_nonr_TI$ASV %in% intersection,]
intersection_r_nonr
```

rPSC vs healthy

```{r}
r_H_TI <- read.xlsx("results/Q2/DAA_final_terminal_ileum.xlsx",sheet = " H vs rPSC Genera")
r_H_colon <- read.xlsx("results/Q2/DAA_final_colon.xlsx",sheet = " H vs rPSC Genera")
intersection <- intersect(r_H_TI$ASV,r_H_colon$ASV)
intersection_r_H <- r_H_TI[r_H_TI$ASV %in% intersection,]
intersection_r_H
```

non-rPSC vs healthy

```{r}
non_H_TI <- read.xlsx("results/Q2/DAA_final_terminal_ileum.xlsx",sheet = " H vs non-rPSC Genera")
non_H_colon <- read.xlsx("results/Q2/DAA_final_colon.xlsx",sheet = " H vs non-rPSC Genera")
intersection <- intersect(non_H_TI$ASV,non_H_colon$ASV)
intersection_nonr_H <- non_H_TI[non_H_TI$ASV %in% intersection,]
intersection_nonr_H
```

Statistics

```{r}
df <- data.frame(`Number of Genera`= c(nrow(intersection_r_nonr),
                               nrow(intersection_r_H),
                               nrow(intersection_nonr_H),
                               nrow(intersection_psc))) 

rownames(df) <- c("rPSC vs non-rPSC", "rPSC vs healthy", "non-rPSC vs healthy","rPSC effect")

df
```

#### ElasticNet - merged cohorts

Summary

```{r}
models_summ_df <- do.call(
  rbind, models_summ[grep("colon",names(models_summ),value = TRUE)])

# save the results
write.xlsx(list(binomial=models_summ_df),"results/Q2/elastic_net_summary_colon.xlsx", rowNames=TRUE)

# see the results
models_summ_df
```

**ROC - ASV level**

```{r}
roc_curve_all_custom(roc_cs[c(9:12)],Q = "Q2")
```

**ROC - Genus level**

```{r}
roc_curve_all_custom(roc_cs[c(13:16)],Q="Q2")
```

# Paper-ready visualizations

## Alpha diversity

**Alpha diversity**

```{r, fig.width=10, fig.height=4, fig.fullwidth=TRUE}
p_A <- alpha_div_plots$`Ileum Custom` + 
  ggtitle("Terminal ileum")+ 
  theme(plot.title = element_text(hjust=0.5,face = "bold",size = 15))

p_B <-  alpha_div_plots$`Colon Custom` + 
  ggtitle("Colon") + 
  theme(plot.title = element_text(hjust=0.5,face = "bold",size = 15))

Q2_alpha <- ggarrange(p_A,ggplot() + theme_minimal(),p_B,nrow=1, ncol=3,
                      widths = c(1,0.1,1))
Q2_alpha
```

## Beta diversity

ASV level

```{r}

```

Genus level

```{r, fig.width=10, fig.height=4, fig.fullwidth=TRUE}
pca_ti <- pca_plots_list$`Q2 Ileum Genus custom` + theme(legend.position = "right")
pca_colon <- pca_plots_list$`Q2 Colon Genus custom` + theme(legend.position = "right")

genus_Q2_beta <- ggarrange(pca_ti,
                           ggplot() + theme_minimal(),
                           pca_colon,ncol=3,
                           widths = c(1,0.1,1))
genus_Q2_beta
```

**Alpha + Beta diversity**

```{r, fig.width=10, fig.height=8, fig.fullwidth=TRUE}
alpha_beta <- ggarrange(Q2_alpha,genus_Q2_beta,
                        ncol = 1,nrow=2,labels = c("A","B"))
alpha_beta
```

## Elastic net


ASV level

```{r}

```

Genus level

```{r, fig.width=10, fig.height=4, fig.fullwidth=TRUE}
roc_cs_ti <- roc_cs[5:8]
roc_cs_colon <- roc_cs[13:16]

roc_curve_ti <- roc_curve_all_custom(roc_cs_ti, Q = "Q2",legend = TRUE)

roc_curve_colon <- roc_curve_all_custom(roc_cs_colon, Q="Q2",legend=TRUE) 

roc_curve_plot <- ggarrange(ggplot() + theme_minimal(),
                            roc_curve_ti,
                            ggplot() + theme_minimal(),
                            roc_curve_colon,
                            ggplot() + theme_minimal(),
                            ncol=5, widths = c(0.1,1,0.2,1,0.1),
                            labels=c("","A","","B",""))

roc_curve_plot <- ggarrange(roc_curve_ti,
                            ggplot() + theme_minimal(),
                            roc_curve_colon,
                            ggplot() + theme_minimal(),
                            ncol=4, widths = c(1,0.1,1,0.1))
roc_curve_plot

```
```{r, fig.width=10, fig.height=12, fig.fullwidth=TRUE}
alpha_beta_elastic <- ggarrange(Q2_alpha,genus_Q2_beta,roc_curve_plot,
                        ncol = 1,nrow=3,labels = LETTERS,heights = c(1,1,1))
alpha_beta_elastic
```

## Dot heatmap - DAA

ASV level

```{r}

```

Genus level

```{r, fig.width=10, fig.height=13, fig.fullwidth=TRUE}
list_for_heatmap_ileum <- c(list_intersections[c(4,5,6)])
list_for_heatmap_colon <- c(list_intersections[c(10,11,12)])

names(list_for_heatmap_colon) <- gsub("Genus","",gsub("colon","",names(list_for_heatmap_colon)))
names(list_for_heatmap_ileum) <- gsub("Genus","",gsub("Ileum","",names(list_for_heatmap_ileum)))

p_ileum <- dot_heatmap_linda(list_for_heatmap_ileum,
                             ileum_taxa_tab) + xlab("") + ylab("") + 
  ggtitle("Terminal ileum") + 
  theme(plot.title = element_text(hjust=0.5,face = "bold",size = 15),
        legend.position = "none")


p_colon <- dot_heatmap_linda(list_for_heatmap_colon,
                             colon_taxa_tab) + xlab("") + ylab("") +
  ggtitle("Colon") + 
  theme(plot.title = element_text(hjust=0.5,face = "bold",size = 15),
        legend.position = "none")

heatmap_plot <- ggarrange(p_ileum,
                          p_colon,
                          ncol = 2) 
heatmap_plot
```

## All together

```{r, fig.width=18, fig.height=10, fig.fullwidth=TRUE}

q2_plot <- ggarrange(alpha_beta_elastic,
                     ggplot()+ theme_minimal(),
                     heatmap_plot,
                      ggplot()+ theme_minimal(),
                     ncol=4,labels=c("","", "D",""),widths = c(1,0.2,0.8,0.15))
q2_plot
```

```{r, eval=FALSE}
pdf("clanok/obrazky/q2/Q2_plot_newalpha.pdf",height = 10,width = 17)
q2_plot
dev.off()
```

# Session info

```{r}
sessionInfo()
```


